---
title: "A workflow for scRNA-seq data analysis: autism dataset from Velmeshev et al, 2019"
author: "Mengqi"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: journal
    highlight: tango
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: true
      smooth_scroll: false
    number_sections: true
editor_options: 
  chunk_output_type: console
---

# Introduction

This markdown is for analyzing the paper [Single-cell genomics identifies cell type–specific molecular changes in autism](https://science.sciencemag.org/content/364/6441/685.abstract). Some details of their analysis is provided [here](https://science.sciencemag.org/content/suppl/2019/05/15/364.6441.685.DC1). Briefly, there analysis are doing individual case-control study using the [MAST](https://www.bioconductor.org/packages/release/bioc/html/MAST.html) packages.

This pipeline using some alternative strategies for data processing and analysis, mostly based on [bioconductor workflows for scRNAseq](https://bioconductor.org/packages/release/workflows/html/simpleSingleCell.html). 

The packages required for the analysis are as follows:
- scater: collection of tools for doing quality control analyses of scRNA-seq
- scran: methods provide normalization of cell-specific biases, correcting batch effects, identify marker genes
- SC3: package for single cell consensus clustering.


```{r global_options, include=FALSE}
knitr::opts_chunk$set(fig.width=8, fig.height=8, fig.path='Figs/',
                      echo=TRUE, warning=FALSE, message=FALSE)
```

# data preprocessing. 

Before running this R markdown, please first download the datafile exprMatrix.tsv.gz and meta.tsv from  Single-cell genomics identifies cell type–specific molecular changes in autism to your local computer, under a folder called Data_PRJNA434002, and unzip the files. Then set another folder as the working directory, beside the folder Data_PRJNA434002. The following code help generate a R-friendly file and some smaller test samples. 

```{r data_preprocessing, warning = FALSE, message = FALSE}

setwd("/fh/fast/sun_w/mengqi/1.Testing_scRNAseq/")

#change the original file into rds for better read-writing-storage
exprM=read.table("../Data_PRJNA434002/exprMatrix.tsv.gz",header=TRUE,row.names=1)

saveRDS(exprM,"../Data_PRJNA434002/exprMatrix.rds")


#generate a 1% percentage cell subfiles, for testing
exprM100=as.matrix(exprM[,seq(1,1045)*100])
saveRDS(exprM100,"../Data_PRJNA434002/exprMatrix100.rds")

meta=read.table("../Data_PRJNA434002/meta.tsv",header = TRUE, sep = "\t")
meta100=meta[seq(1,1045)*100,]
saveRDS(meta100,"../Data_PRJNA434002/meta100.rds")

#generate another 10% percentage cells with first 3000 genes, for testing
meta10=meta[seq(1,10455)*10,]
saveRDS(meta10,"../Data_PRJNA434002/meta10.rds")
exprM3k10=exprM[1:3000,seq(1,10455)*10]
saveRDS(exprM3k10,"../Data_PRJNA434002/exprMatrix3k10.rds")
exprM1k10=exprM[1:1000,seq(1,10455)*10]
saveRDS(exprM1k10,"../Data_PRJNA434002/exprMatrix1k10.rds")
```

# tSNE plots

We repeated the tSNE plots for cells based on the truncated SVDs(k=50).According to [Velmeshev et al,2019](https://science.sciencemag.org/content/suppl/2019/05/15/364.6441.685.DC1),to avoid batch effect, we remove the features who has the rsquared >0.2 with any batch features through the regression (In the original paper, they remove features have more than 0.2 pearson correlation with batch effect). 

```{r tSNE_plot, fig.dim=c(8,8)}
library("RSpectra")
library("Rtsne")

exprM=as.matrix(readRDS("../Data_PRJNA434002/exprMatrix.rds"))
meta=read.table("../Data_PRJNA434002/meta.tsv",header = TRUE, sep = "\t")

k = 50
svd50=svds(exprM, k)
#svd50=readRDS("../Data_PRJNA434002/svd50.rds")

plot(svd50$d^2/sum(svd50$d^2), xlim = c(0, k),ylim=c(0,0.02), type = "b", pch = 16, xlab = "principal components", 
     ylab = "variance explained")
k=16
svd50v=svd50$v[,1:k]

xc=as.matrix(meta[,"Capbatch"])
xs=as.matrix(meta[,"Seqbatch"])
cor_c=matrix(ncol=1,nrow=k)
cor_s=matrix(ncol=1,nrow=k)

for(ik in 1:k){
  y=svd50v[,ik]
  ##calculate correlation of capbatch
  lmc = lm(y ~ as.factor(xc))
  cor_c[ik]=summary(lmc)$r.square
  ##calculate correlation of seqbatch
  lms = lm(y ~ as.factor(xs))
  cor_s[ik]=sqrt(summary(lms)$r.square)
  
}


#cor 0.1 threshold 
flag=(abs(cor_c)<0.1 & abs(cor_s)<0.1) 
svd50v0.1=svd50v[,flag==1]
tsne=Rtsne(svd50v0.1,dims=2, perplexity=ncol(svd50v0.1))

pdf("tSNE_plots_cor0.1.pdf",height = 8,width = 8)
plot(tsne$Y, cex=.1,main="tSNE-Capbatch",col=as.numeric(as.factor(meta[,"Capbatch"])))
plot(tsne$Y, cex=.1,main="tSNE-Seqbatch",col=as.numeric(as.factor(meta[,"Seqbatch"])))
plot(tsne$Y, cex=.2,main="tSNE-cluster",col=as.numeric(as.factor(meta[,"cluster"])))
plot(tsne$Y, cex=.2,main="tSNE-individual",col=as.numeric(as.factor(meta[,"individual"])))
dev.off()

#cor 0.2 threshold 
flag=(abs(cor_c)<0.2 & abs(cor_s)<0.2) 
svd50v0.2=svd50$v0.2[,flag==1]
tsne=Rtsne(svd50v0.2,dims=2, perplexity=ncol(svd50v0.2))

pdf("tSNE_plots_cor0.2.pdf",height = 8,width = 8)
plot(tsne$Y, cex=.1,main="tSNE-Capbatch",col=as.numeric(as.factor(meta[,"Capbatch"])))
plot(tsne$Y, cex=.1,main="tSNE-Seqbatch",col=as.numeric(as.factor(meta[,"Seqbatch"])))
plot(tsne$Y, cex=.2,main="tSNE-cluster",col=as.numeric(as.factor(meta[,"cluster"])))
plot(tsne$Y, cex=.2,main="tSNE-individual",col=as.numeric(as.factor(meta[,"individual"])))
dev.off()

```


# MAST and permutation results

To detect the type I error controls of the individual level case-control method, we apply the MAST regresion steps and permutated the diagnosis information among the individuals. To do so, we:

## Establish the sca files from the matrix

For each of the 17 clusters, we do the analysis separately.

```{r establish_sca}
#something from the header file
cluster_tag=1 #this tag indicate the clusters it can be choose in 1 to 17
perm_tag=0    #this tage indicate the permutation tags, 0 means no permutation, otherwise, permutation id

library("MAST")
library("lme4")

meta=readRDS("../Data_PRJNA434002/meta10.rds")
exprM=readRDS("../Data_PRJNA434002/exprMatrix3k10.rds")
cur_cluster=as.character(unique(meta$cluster)[cluster_tag])
exprM=as.matrix(exprM[,meta$cluster==cur_cluster])
meta=meta[meta$cluster==cur_cluster,]

dim(exprM)
exprM[1:10,1:10]
cell_list=colnames(exprM)
gene_list=rownames(exprM)
rownames(exprM)=gene_list

fData=data.frame(primerid=gene_list)

cData=data.frame(wellKey=cell_list)
colnames(meta)
length(fData)
length(cData)

sca=FromMatrix(exprM, cData, fData)
colData(sca)$cngeneson = as.numeric((colSums(exprM > 0))/nrow(exprM)) #from Chong and Paul
colData(sca)$diagnosis = meta$diagnosis
colData(sca)$ind = as.factor(meta$individual)
colData(sca)$age = as.numeric(meta$age)
colData(sca)$sex = as.factor(meta$sex)
colData(sca)$RIN = as.numeric(meta$RNA.Integrity.Number)
colData(sca)$PMI = as.numeric(meta$post.mortem.interval..hours.)
colData(sca)$region = as.factor(meta$region)
colData(sca)$Capbatch = as.factor(meta$Capbatch)
colData(sca)$Seqbatch = as.factor(meta$Seqbatch)
colData(sca)$ribo_perc = as.numeric(meta$RNA.ribosomal.percent)

colData(sca)
```

## Do the permutation

If permutation needed, we do the permutation by mismatch the cases and controls from the individual diagnosis information.
```{r MAST_and_permutation}
if(perm_tag>0){
  #count cases and controls
  diag_info=paste0(colData(sca)$ind,":",colData(sca)$diagnosis)
  diag_kind=unique(diag_info)
  diag_kind=t(apply(as.matrix(diag_kind),1,function(x){return(unlist(strsplit(x,":")))}))
  
  #permute
  diag_kind[,2]=diag_kind[sample.int(nrow(diag_kind),nrow(diag_kind),replace=F),2]
  
  #match back to each individuals
  ind_index=match(colData(sca)$ind,diag_kind[,1])
  colData(sca)$diagnosis=as.factor(diag_kind[ind_index,2])
}
```


## Apply the MAST
We fit the model according to [Velmeshev et al,2019](https://science.sciencemag.org/content/suppl/2019/05/15/364.6441.685.DC1),
zlm(~diagnosis + (1|ind) + cngeneson + age + sex + RIN + PMI + region + Capbatch + Seqbatch 
+ ribo_perc, sca, method = "glmer", ebayes = F, silent=T).


```{r apply_MAST,results="hide"}
b=zlm(formula=~diagnosis + ( 1 | ind ) + cngeneson + age + sex + RIN + PMI + region + Capbatch + Seqbatch 
      + ribo_perc, sca=sca, method = "glmer", ebayes = F, silent=T)
#saveRDS(b, paste0("../Data_PRJNA434002/zlm_3k10_",cluster_tag,"_",perm_tag,".rds"))
```


## Summary the fit Results
```{r fit_result_summary}
bs=summary(b,logFC=TRUE,doLRT = paste0("diagnosis","Control"), level = 0.95)
#saveRDS(bs, paste0("../Data_PRJNA434002/zlms_3k10_",cluster_tag,"_",perm_tag,".rds"))
bs$datatable

```

# Session information
```{r}
sessionInfo()
```

# Reference
