<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>Introduction</title>

<script type="text/javascript">
window.onload = function() {
  var imgs = document.getElementsByTagName('img'), i, img;
  for (i = 0; i < imgs.length; i++) {
    img = imgs[i];
    // center an image if it is the only element of its parent
    if (img.parentElement.childElementCount === 1)
      img.parentElement.style.textAlign = 'center';
  }
};
</script>





<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 13px;
}

body {
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 20px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 {
   font-size:2.2em;
}

h2 {
   font-size:1.8em;
}

h3 {
   font-size:1.4em;
}

h4 {
   font-size:1.0em;
}

h5 {
   font-size:0.9em;
}

h6 {
   font-size:0.8em;
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre, img {
  max-width: 100%;
}
pre {
  overflow-x: auto;
}
pre code {
   display: block; padding: 0.5em;
}

code {
  font-size: 92%;
  border: 1px solid #ccc;
}

code[class] {
  background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * {
      background: transparent !important;
      color: black !important;
      filter:none !important;
      -ms-filter: none !important;
   }

   body {
      font-size:12pt;
      max-width:100%;
   }

   a, a:visited {
      text-decoration: underline;
   }

   hr {
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote {
      padding-right: 1em;
      page-break-inside: avoid;
   }

   tr, img {
      page-break-inside: avoid;
   }

   img {
      max-width: 100% !important;
   }

   @page :left {
      margin: 15mm 20mm 15mm 10mm;
   }

   @page :right {
      margin: 15mm 10mm 15mm 20mm;
   }

   p, h2, h3 {
      orphans: 3; widows: 3;
   }

   h2, h3 {
      page-break-after: avoid;
   }
}
</style>



</head>

<body>
<h1>Introduction</h1>

<p>This markdown is for analyzing the <a href="https://www.nature.com/articles/nmeth.4407">Habib et al. 2017 DroNc dataset</a>.</p>

<h1>Obtaining/Loading Counts</h1>

<p>The dataset is available <a href="https://storage.googleapis.com/gtex_additional_datasets/single_cell_data/GTEx_droncseq_hip_pcf.tar">here</a>.</p>

<pre><code class="r">where_file = c(&quot;mycomp&quot;,&quot;longleaf&quot;,&quot;remote&quot;)[2]
if(where_file == &quot;mycomp&quot;){
  drive = strsplit(getwd(),&quot;/&quot;)[[1]][1]
  src_dir = file.path(drive,&quot;Files to KEEP/SOURCE_files&quot;)
  data_dir = file.path(drive,&quot;Files to KEEP/BIOS_eQTL/documents/dronc_seq&quot;)
} else if(where_file == &quot;longleaf&quot;){
  data_dir = &quot;/pine/scr/p/l/pllittle/CS_eQTL/s3_Real/original/DroNc/GTEx_droncseq_hip_pcf&quot;
  src_dir = system(&quot;echo $HOME&quot;,intern = TRUE)
} else if(where_file == &quot;remote&quot;){
  data_dir = &quot;C:/Users/pllittle/Downloads/dronc_seq&quot;
}

# Source/Libraries
source(&quot;https://bioconductor.org/biocLite.R&quot;)
</code></pre>

<pre><code>## Bioconductor version 3.7 (BiocInstaller 1.30.0), ?biocLite for help
</code></pre>

<pre><code class="r">source(file.path(src_dir,&quot;SOURCE.R&quot;))
# biocLite(&quot;BiocUpgrade&quot;)
bio_packs = c(&quot;SingleCellExperiment&quot;,&quot;DropletUtils&quot;,&quot;biomaRt&quot;,
              &quot;scater&quot;,&quot;scran&quot;,&quot;limma&quot;,&quot;org.Hs.eg.db&quot;,&quot;SC3&quot;)
if( !all(bio_packs %in% installed.packages()[,&quot;Package&quot;]) ){
  biocLite(bio_packs,suppressUpdates = TRUE)
  # biocLite(&quot;SC3&quot;,suppressUpdates = TRUE)
}
cran_packs = c(&quot;stringi&quot;,&quot;irlba&quot;)
if( !all(cran_packs %in% installed.packages()[,&quot;Package&quot;]) ){
  install.packages(cran_packs)
}
# DropletUtils: provides functions for data from droplet technologies such as 10X Genomics
# biomaRt: provides easy access to databases, such as Ensembl, COSMIC, Uniprot, HGNC, etc.
# scater: collection of tools for doing quality control analyses of scRNA-seq
# scran: methods provide normalization of cell-specific biases, correcting batch effects, identify marker genes
suppressPackageStartupMessages(library(SingleCellExperiment))
suppressPackageStartupMessages(library(DropletUtils))
suppressPackageStartupMessages(library(biomaRt))
suppressPackageStartupMessages(library(scater))
suppressPackageStartupMessages(library(scran))
suppressPackageStartupMessages(library(limma))

# Import counts
counts_fn = file.path(data_dir,&quot;image.rds&quot;)
if( !file.exists(counts_fn) ){
  num_lines = 32112
  counts = read.delim(file.path(data_dir,&quot;GTEx_droncseq_hip_pcf.umi_counts.txt&quot;),
    sep=&#39;\t&#39;,header=TRUE,row.names = 1,check.names = FALSE,nrows = num_lines-1)
  print(dim(counts))
  saveRDS(counts,counts_fn)
}
counts = readRDS(counts_fn)
sce = SingleCellExperiment(assays = list(counts = as.matrix(counts)))
rm(counts)
sce
</code></pre>

<pre><code>## class: SingleCellExperiment 
## dim: 32111 14963 
## metadata(0):
## assays(1): counts
## rownames(32111): A1BG A1BG-AS1 ... ZZEF1 ZZZ3
## rowData names(0):
## colnames(14963): hHP1_AACACTATCTAC hHP1_CTACGCATCCAT ...
##   PFC-CD_CTCCATTCATGC PFC-CD_CGTCATTAGCAG
## colData names(0):
## reducedDimNames(0):
## spikeNames(0):
</code></pre>

<pre><code class="r"># Define spikes
grep(&quot;^ERCC&quot;,rownames(sce))
</code></pre>

<pre><code>## [1] 8566 8567 8568 8569 8570 8571 8572 8573 8574
</code></pre>

<pre><code class="r">rownames(sce)[grep(&quot;^ERCC&quot;,rownames(sce))]
</code></pre>

<pre><code>## [1] &quot;ERCC1&quot;   &quot;ERCC2&quot;   &quot;ERCC3&quot;   &quot;ERCC4&quot;   &quot;ERCC5&quot;   &quot;ERCC6&quot;   &quot;ERCC6L&quot; 
## [8] &quot;ERCC6L2&quot; &quot;ERCC8&quot;
</code></pre>

<pre><code class="r"># isSpike(sce,&quot;ERCC&quot;) = grepl(&quot;^ERCC&quot;,rownames(sce))
# sce = calculateQCMetrics(sce,feature_control=list(ERCC=is_spike))
</code></pre>

<h1>Pre-processing: Quality Control, Gene Detection, Normalization</h1>

<h2>Gene Annotation</h2>

<pre><code class="r"># Code from Dr. Sun&#39;s Rmd
anno.file = &quot;~/research/scRNAseq/workflow/data/gene.annoation.rds&quot;
if(file.exists(anno.file)){
  gene.annotation = readRDS(anno.file)
}else{
  ensembl = useMart(&quot;ensembl&quot;,dataset=&quot;hsapiens_gene_ensembl&quot;)

  attr.string = c(&#39;ensembl_gene_id&#39;, &#39;hgnc_symbol&#39;, &#39;chromosome_name&#39;)
  attr.string = c(attr.string, &#39;start_position&#39;, &#39;end_position&#39;, &#39;strand&#39;)
  attr.string = c(attr.string, &#39;description&#39;, &#39;percentage_gene_gc_content&#39;)
  attr.string = c(attr.string, &#39;gene_biotype&#39;)

  rowData(sce)[1:2,]
  gene.annotation = getBM(attributes=attr.string, 
                          filters =  &#39;ensembl_gene_id&#39;, 
                          values = rowData(sce)$ID, 
                          mart = ensembl)
}

dim(gene.annotation)
</code></pre>

<pre><code class="r"># My attempt to annotate gene symbols with ensembl
# ensembl = useMart(&quot;ensembl&quot;,dataset=&quot;hsapiens_gene_ensembl&quot;)
ensembl = useEnsembl(biomart=&quot;ensembl&quot;,GRCh=37,dataset=&quot;hsapiens_gene_ensembl&quot;)
attr_string = c(&quot;hgnc_symbol&quot;,&quot;external_gene_name&quot;,&quot;chromosome_name&quot;,
            &quot;start_position&quot;,&quot;end_position&quot;,&quot;strand&quot;,&quot;description&quot;,
            &quot;percentage_gene_gc_content&quot;,&quot;gene_biotype&quot;)
gene_anno = getBM(attributes = attr_string,
                  filters = &quot;external_gene_name&quot;,
                  values = rownames(sce),
                  mart = ensembl)

# Check the sources of annotation discrepancy
dim(sce); dim(gene_anno)
</code></pre>

<pre><code>## [1] 32111 14963
</code></pre>

<pre><code>## [1] 34526     9
</code></pre>

<pre><code class="r">gene_missing = setdiff(rownames(sce),gene_anno$external_gene_name)
gene_missing[1:10]
</code></pre>

<pre><code>##  [1] &quot;AC005152.2&quot; &quot;AC006132.1&quot; &quot;AC006445.8&quot; &quot;AC007092.1&quot; &quot;AC007390.5&quot;
##  [6] &quot;AC007464.1&quot; &quot;AC009232.2&quot; &quot;AC009236.1&quot; &quot;AC010127.5&quot; &quot;AC012307.3&quot;
</code></pre>

<pre><code class="r">length(gene_missing)
</code></pre>

<pre><code>## [1] 157
</code></pre>

<pre><code class="r">gene_anno[which(!(gene_anno$external_gene_name %in% rownames(sce))),]
</code></pre>

<pre><code>##      hgnc_symbol external_gene_name chromosome_name start_position
## 4232                       C10ORF68              10       32832297
## 4645                       C1ORF220               1      178511950
##      end_position strand
## 4232     33171802      1
## 4645    178517579      1
##                                                                                                               description
## 4232 Homo sapiens coiled-coil domain containing 7 (CCDC7), transcript variant 5, mRNA. [Source:RefSeq mRNA;Acc:NM_024688]
## 4645                                                                                                                     
##      percentage_gene_gc_content   gene_biotype
## 4232                      37.23 protein_coding
## 4645                      49.91 protein_coding
</code></pre>

<pre><code class="r">gene_anno = gene_anno[which(gene_anno$external_gene_name %in% rownames(sce)),]
dim(sce); dim(gene_anno)
</code></pre>

<pre><code>## [1] 32111 14963
</code></pre>

<pre><code>## [1] 34524     9
</code></pre>

<pre><code class="r">table(gene_anno$chromosome_name)
</code></pre>

<pre><code>## 
##                        1                       10                       11 
##                     3084                     1252                     1741 
##                       12                       13                       14 
##                     1694                      637                     1145 
##                       15                       16                       17 
##                     1169                     1426                     1820 
##                       18                       19                        2 
##                      633                     1955                     2186 
##                       20                       21                       22 
##                      800                      383                      704 
##                        3                        4                        5 
##                     1784                     1301                     1567 
##                        6                        7                        8 
##                     1622                     1446                     1296 
##                        9               GL000192.1               GL000193.1 
##                     1207                        2                        1 
##               GL000194.1               GL000195.1               GL000199.1 
##                        2                        1                        1 
##               GL000204.1               GL000205.1               GL000213.1 
##                        1                        2                        1 
##               GL000218.1               GL000219.1               GL000220.1 
##                        1                        2                        3 
##               GL000223.1               GL000237.1               GL000241.1 
##                        1                        1                        1 
##               GL000242.1             HG1032_PATCH        HG104_HG975_PATCH 
##                        1                        3                       22 
##             HG1063_PATCH             HG1074_PATCH             HG1079_PATCH 
##                        3                        4                       43 
##       HG1082_HG167_PATCH             HG1091_PATCH             HG1133_PATCH 
##                       33                        1                       15 
##              HG115_PATCH             HG1208_PATCH             HG1211_PATCH 
##                        4                        3                       19 
##              HG122_PATCH             HG1257_PATCH             HG1287_PATCH 
##                       18                       51                       95 
##             HG1292_PATCH             HG1293_PATCH             HG1304_PATCH 
##                       15                       11                        9 
##             HG1308_PATCH             HG1322_PATCH       HG1350_HG959_PATCH 
##                       18                        1                       16 
##               HG14_PATCH   HG142_HG150_NOVEL_TEST             HG1423_PATCH 
##                        1                        1                        3 
##             HG1424_PATCH             HG1425_PATCH             HG1426_PATCH 
##                        4                        4                       10 
##             HG1433_PATCH             HG1434_PATCH             HG1435_PATCH 
##                       34                        1                        3 
##      HG1436_HG1432_PATCH             HG1437_PATCH             HG1438_PATCH 
##                       55                        5                        9 
##             HG1439_PATCH              HG144_PATCH             HG1440_PATCH 
##                       15                        1                        5 
##             HG1441_PATCH             HG1442_PATCH      HG1443_HG1444_PATCH 
##                        6                        5                       10 
##             HG1453_PATCH             HG1458_PATCH             HG1459_PATCH 
##                        5                        4                       25 
##             HG1462_PATCH             HG1463_PATCH             HG1479_PATCH 
##                       10                        6                        1 
##             HG1497_PATCH             HG1500_PATCH             HG1501_PATCH 
##                       75                        2                        1 
##             HG1502_PATCH             HG1592_PATCH             HG1595_PATCH 
##                        3                       14                        8 
##             HG1699_PATCH        HG174_HG254_PATCH              HG183_PATCH 
##                        1                        6                        6 
##              HG185_PATCH              HG186_PATCH               HG19_PATCH 
##                       19                        5                        4 
##              HG193_PATCH              HG237_PATCH              HG243_PATCH 
##                        2                        5                       14 
##              HG256_PATCH               HG27_PATCH              HG271_PATCH 
##                        3                        4                        6 
##              HG280_PATCH               HG29_PATCH              HG299_PATCH 
##                        1                       12                       13 
##              HG305_PATCH              HG306_PATCH              HG311_PATCH 
##                        1                        5                        3 
##              HG325_PATCH              HG329_PATCH              HG339_PATCH 
##                       13                        5                        2 
##              HG344_PATCH              HG348_PATCH              HG357_PATCH 
##                       19                        5                        6 
##              HG375_PATCH              HG385_PATCH        HG388_HG400_PATCH 
##                        1                       15                       13 
##              HG414_PATCH              HG417_PATCH              HG418_PATCH 
##                        4                        3                        4 
##              HG444_PATCH        HG480_HG481_PATCH              HG497_PATCH 
##                        1                        1                        1 
##               HG50_PATCH HG506_HG507_HG1000_PATCH              HG531_PATCH 
##                        2                        3                        1 
##              HG536_PATCH              HG544_PATCH              HG686_PATCH 
##                        3                        5                        1 
##                HG7_PATCH              HG706_PATCH              HG729_PATCH 
##                       21                        1                        4 
##              HG730_PATCH              HG736_PATCH              HG745_PATCH 
##                        7                        1                        8 
##              HG747_PATCH              HG748_PATCH               HG75_PATCH 
##                       10                        5                       10 
##               HG79_PATCH              HG858_PATCH              HG865_PATCH 
##                       12                        5                        4 
##              HG871_PATCH              HG873_PATCH              HG883_PATCH 
##                        8                        3                       14 
##              HG905_PATCH              HG944_PATCH              HG946_PATCH 
##                        2                        1                        3 
##              HG953_PATCH              HG957_PATCH              HG962_PATCH 
##                       16                        3                        6 
##              HG971_PATCH              HG979_PATCH              HG987_PATCH 
##                        5                        1                        6 
##              HG989_PATCH              HG990_PATCH              HG991_PATCH 
##                        2                        3                        1 
##              HG996_PATCH            HG998_1_PATCH            HG998_2_PATCH 
##                        3                        2                        1 
##            HG999_1_PATCH           HSCHR1_1_CTG31           HSCHR1_2_CTG31 
##                        1                        3                        6 
##           HSCHR1_3_CTG31           HSCHR10_1_CTG2           HSCHR10_1_CTG5 
##                        2                        1                        5 
##           HSCHR12_1_CTG1         HSCHR12_1_CTG2_1           HSCHR12_1_CTG5 
##                        6                        3                        3 
##           HSCHR12_2_CTG2         HSCHR12_2_CTG2_1         HSCHR12_3_CTG2_1 
##                       16                        1                        2 
##           HSCHR15_1_CTG4           HSCHR15_1_CTG8         HSCHR16_1_CTG3_1 
##                        2                        1                        1 
##         HSCHR16_2_CTG3_1                HSCHR17_1           HSCHR17_1_CTG1 
##                        2                        3                        3 
##           HSCHR17_1_CTG4           HSCHR17_3_CTG4           HSCHR17_4_CTG4 
##                        2                        1                        8 
##           HSCHR17_6_CTG4         HSCHR18_1_CTG1_1           HSCHR18_2_CTG2 
##                        1                        2                        4 
##         HSCHR18_2_CTG2_1           HSCHR19_1_CTG3         HSCHR19_1_CTG3_1 
##                        3                        3                        2 
##           HSCHR19_2_CTG3           HSCHR19_3_CTG3     HSCHR19LRC_COX1_CTG1 
##                        3                        3                       36 
##     HSCHR19LRC_COX2_CTG1    HSCHR19LRC_LRC_I_CTG1    HSCHR19LRC_LRC_J_CTG1 
##                       29                       24                       24 
##    HSCHR19LRC_LRC_S_CTG1    HSCHR19LRC_LRC_T_CTG1     HSCHR19LRC_PGF1_CTG1 
##                       24                       24                       43 
##     HSCHR19LRC_PGF2_CTG1            HSCHR2_1_CTG1           HSCHR2_1_CTG12 
##                       30                        2                        4 
##           HSCHR2_2_CTG12           HSCHR20_1_CTG1         HSCHR21_2_CTG1_1 
##                        1                        1                        1 
##         HSCHR21_3_CTG1_1         HSCHR21_4_CTG1_1           HSCHR22_1_CTG1 
##                        3                        5                        6 
##           HSCHR22_1_CTG2           HSCHR22_2_CTG1            HSCHR3_1_CTG1 
##                        2                        7                        2 
##          HSCHR3_1_CTG2_1                 HSCHR4_1           HSCHR4_1_CTG12 
##                        2                        3                        2 
##            HSCHR4_1_CTG6            HSCHR4_2_CTG9            HSCHR5_1_CTG1 
##                        1                        1                       14 
##            HSCHR5_1_CTG2            HSCHR5_1_CTG5            HSCHR5_2_CTG1 
##                        1                        1                        1 
##            HSCHR5_3_CTG1            HSCHR6_1_CTG5           HSCHR6_MHC_APD 
##                        1                        1                       89 
##           HSCHR6_MHC_COX           HSCHR6_MHC_DBB          HSCHR6_MHC_MANN 
##                      173                      161                      140 
##           HSCHR6_MHC_MCF           HSCHR6_MHC_QBL          HSCHR6_MHC_SSTO 
##                      156                      166                      146 
##            HSCHR7_1_CTG6           HSCHR9_2_CTG35                       MT 
##                        7                        2                       33 
##                        X                        Y 
##                     1047                       46
</code></pre>

<pre><code class="r"># gene_anno = gene_anno[which(gene_anno$chromosome_name %in% c(1:22,&quot;X&quot;,&quot;Y&quot;,&quot;MT&quot;)),]
# dim(sce); dim(gene_anno)
t1 = table(gene_anno$external_gene_name)
t2 = t1[t1 &gt; 1]
length(t2)
</code></pre>

<pre><code>## [1] 1448
</code></pre>

<pre><code class="r">t2[1:10]
</code></pre>

<pre><code>## 
##     AADACL2 AB019440.50      ABCB11       ABCD1       ABCF1      ABHD11 
##           2           2           2           2           7           2 
##     ABHD16A         ABO        ABT1  AC003958.2 
##           7           2           2           2
</code></pre>

<pre><code class="r">gene_anno[which(gene_anno$external_gene_name %in% names(t2)[1:5]),]
</code></pre>

<pre><code>##     hgnc_symbol external_gene_name chromosome_name start_position
## 48      AADACL2            AADACL2 HSCHR3_1_CTG2_1      151462242
## 167       ABCF1              ABCF1  HSCHR6_MHC_DBB       30529391
## 205       ABCF1              ABCF1  HSCHR6_MHC_QBL       30528662
## 305                    AB019440.50              14      106663246
## 334       ABCF1              ABCF1 HSCHR6_MHC_MANN       30583828
## 418       ABCF1              ABCF1 HSCHR6_MHC_SSTO       30530573
## 419       ABCF1              ABCF1  HSCHR6_MHC_MCF       30617610
## 421       ABCF1              ABCF1               6       30539153
## 431      ABCB11             ABCB11  HSCHR2_1_CTG12      169796437
## 445     AADACL2            AADACL2               3      151451704
## 458      ABCB11             ABCB11               2      169779448
## 501       ABCD1              ABCD1               X      152990323
## 507       ABCF1              ABCF1  HSCHR6_MHC_COX       30529000
## 512                    AB019440.50    HG1592_PATCH      106672940
## 522       ABCD1              ABCD1    HG1497_PATCH      152893340
##     end_position strand
## 48     151489665      1
## 167     30555197      1
## 205     30554466      1
## 305    106665214      1
## 334     30609632      1
## 418     30556377      1
## 419     30643414      1
## 421     30564956      1
## 431    169904821     -1
## 445    151479127      1
## 458    169887832     -1
## 501    153010216      1
## 507     30554805      1
## 512    106674908      1
## 522    152913233      1
##                                                                             description
## 48                      arylacetamide deacetylase-like 2 [Source:HGNC Symbol;Acc:24427]
## 167    ATP-binding cassette, sub-family F (GCN20), member 1 [Source:HGNC Symbol;Acc:70]
## 205    ATP-binding cassette, sub-family F (GCN20), member 1 [Source:HGNC Symbol;Acc:70]
## 305                                                                                    
## 334    ATP-binding cassette, sub-family F (GCN20), member 1 [Source:HGNC Symbol;Acc:70]
## 418    ATP-binding cassette, sub-family F (GCN20), member 1 [Source:HGNC Symbol;Acc:70]
## 419    ATP-binding cassette, sub-family F (GCN20), member 1 [Source:HGNC Symbol;Acc:70]
## 421    ATP-binding cassette, sub-family F (GCN20), member 1 [Source:HGNC Symbol;Acc:70]
## 431 ATP-binding cassette, sub-family B (MDR/TAP), member 11 [Source:HGNC Symbol;Acc:42]
## 445                     arylacetamide deacetylase-like 2 [Source:HGNC Symbol;Acc:24427]
## 458 ATP-binding cassette, sub-family B (MDR/TAP), member 11 [Source:HGNC Symbol;Acc:42]
## 501      ATP-binding cassette, sub-family D (ALD), member 1 [Source:HGNC Symbol;Acc:61]
## 507    ATP-binding cassette, sub-family F (GCN20), member 1 [Source:HGNC Symbol;Acc:70]
## 512                                                                                    
## 522      ATP-binding cassette, sub-family D (ALD), member 1 [Source:HGNC Symbol;Acc:61]
##     percentage_gene_gc_content   gene_biotype
## 48                       35.30 protein_coding
## 167                      48.02 protein_coding
## 205                      47.93 protein_coding
## 305                      46.47        lincRNA
## 334                      47.96 protein_coding
## 418                      47.93 protein_coding
## 419                      47.96 protein_coding
## 421                      47.97 protein_coding
## 431                      36.93 protein_coding
## 445                      35.30 protein_coding
## 458                      36.93 protein_coding
## 501                      58.36 protein_coding
## 507                      47.93 protein_coding
## 512                      46.22        lincRNA
## 522                      58.36 protein_coding
</code></pre>

<pre><code class="r">gene_anno = dplyr::distinct(gene_anno,external_gene_name,.keep_all = TRUE)

if(FALSE){ # Old code
t1 = table(gene_anno$hgnc_symbol)
t2 = t1[t1 &gt; 1]
t2
gene_anno[which(gene_anno$hgnc_symbol %in% names(t2)),]
gene_anno = dplyr::distinct(gene_anno,hgnc_symbol,.keep_all = TRUE)
}

dim(gene_anno)
</code></pre>

<pre><code>## [1] 31954     9
</code></pre>

<pre><code class="r">table(gene_anno$gene_biotype)
</code></pre>

<pre><code>## 
## 3prime_overlapping_ncrna                antisense                IG_C_gene 
##                       11                     4136                        3 
##                IG_V_gene                  lincRNA                    miRNA 
##                        6                     4790                      972 
##                 misc_RNA                  Mt_rRNA                  Mt_tRNA 
##                      935                        2                       18 
##     processed_transcript           protein_coding               pseudogene 
##                      436                    18301                        4 
##                     rRNA           sense_intronic        sense_overlapping 
##                      206                      648                      172 
##                   snoRNA                    snRNA                TR_C_gene 
##                      202                     1083                        5 
##                TR_J_gene                TR_V_gene 
##                        7                       17
</code></pre>

<pre><code class="r">gene_missing = setdiff(rownames(sce),gene_anno$external_gene_name)
gene_missing[1:10]
</code></pre>

<pre><code>##  [1] &quot;AC005152.2&quot; &quot;AC006132.1&quot; &quot;AC006445.8&quot; &quot;AC007092.1&quot; &quot;AC007390.5&quot;
##  [6] &quot;AC007464.1&quot; &quot;AC009232.2&quot; &quot;AC009236.1&quot; &quot;AC010127.5&quot; &quot;AC012307.3&quot;
</code></pre>

<pre><code class="r">length(gene_missing)
</code></pre>

<pre><code>## [1] 157
</code></pre>

<pre><code class="r"># Subset out genes without annotation
w2kp = match(gene_anno$external_gene_name,rownames(sce))
any(is.na(w2kp))
</code></pre>

<pre><code>## [1] FALSE
</code></pre>

<pre><code class="r">gene_anno$external_gene_name[which(is.na(w2kp))]
</code></pre>

<pre><code>## character(0)
</code></pre>

<pre><code class="r">sce = sce[w2kp,]
dim(sce)
</code></pre>

<pre><code>## [1] 31954 14963
</code></pre>

<pre><code class="r">rowData(sce) = gene_anno
</code></pre>

<h2>Identify Low quality cells</h2>

<h3>barcodeRanks filtering</h3>

<p>Look <a href="https://bioconductor.org/packages/release/workflows/vignettes/simpleSingleCell/inst/doc/work-3-tenx.html#calling-cells-from-empty-droplets">here</a> for reference.</p>

<!--- From the above analysis, some cells with very small number of UMI's may also have small FDR suggesting the distribution of UMI counts are different from what is expected from ambient profile. We choose a more conservative strategy, to keep the cells with total number of UMI larger than the inflection point estimate (bcrank\$inflection=```bcrank$inflection```) and FDR < 0.01.) -->

<h3>Incorporate Mito and Ribo to calculate QC metrics</h3>

<p>Next step we apply more QC based on a set of features per cell. We&#39;ll look at ribosomal genes. The imported file can be found <a href="https://www.genenames.org/cgi-bin/genefamilies/set/1054/download/branch">here</a>.</p>

<pre><code class="r">file_link = &quot;https://www.genenames.org/cgi-bin/genefamilies/set/1054/download/branch&quot;
file_name = strsplit(file_link,&quot;/&quot;)[[1]]
file_name = file_name[length(file_name)]
ribo_fn = file.path(data_dir,file_name)
if( !file.exists(ribo_fn) ){
  cmd = sprintf(&quot;cd %s; wget %s&quot;,data_dir,file_link)
  print(cmd)
  system(cmd)
}

ribo = read.table(ribo_fn,sep=&#39;\t&#39;,header=TRUE,stringsAsFactors = FALSE)
ribo[1:2,]
</code></pre>

<pre><code>##   HGNC.ID Approved.Symbol          Approved.Name   Status Previous.Symbols
## 1   10298           RPL10  ribosomal protein L10 Approved                 
## 2   10299          RPL10A ribosomal protein L10a Approved            NEDD6
##                                  Synonyms Chromosome Accession.Numbers
## 1 NOV, QM, DXS648E, DXS648, FLJ23544, L10       Xq28          AB007170
## 2                            Csa-19, L10A    6p21.31            U12404
##   RefSeq.IDs Gene.Family.Tag Gene.family.description Gene.family.ID
## 1  NM_006013             RPL    L ribosomal proteins            729
## 2  NM_007104             RPL    L ribosomal proteins            729
</code></pre>

<pre><code class="r"># Temp save image
post_ribo_download_fn = file.path(data_dir,&quot;post_ribo_download.rds&quot;)
if(where_file == &quot;longleaf&quot;){
  saveRDS(list(sce=sce,ribo=ribo),post_ribo_download_fn)
} else if(FALSE){
  rds = readRDS(post_ribo_download_fn)
  sce = rds$sce
  ribo = rds$ribo
  rm(rds)
}

is_mito = which(rowData(sce)$chromosome_name == &quot;MT&quot;)
is_ribo = which(rowData(sce)$hgnc_symbol %in% ribo$Approved.Symbol)
length(is_mito)
</code></pre>

<pre><code>## [1] 33
</code></pre>

<pre><code class="r">length(is_ribo)
</code></pre>

<pre><code>## [1] 160
</code></pre>

<pre><code class="r">sce = calculateQCMetrics(sce, feature_controls=list(Mt=is_mito, Ri=is_ribo))
sort(colnames(colData(sce)))
</code></pre>

<pre><code>##  [1] &quot;is_cell_control&quot;                               
##  [2] &quot;log10_total_counts&quot;                            
##  [3] &quot;log10_total_counts_endogenous&quot;                 
##  [4] &quot;log10_total_counts_feature_control&quot;            
##  [5] &quot;log10_total_counts_Mt&quot;                         
##  [6] &quot;log10_total_counts_Ri&quot;                         
##  [7] &quot;log10_total_features&quot;                          
##  [8] &quot;log10_total_features_by_counts&quot;                
##  [9] &quot;log10_total_features_by_counts_endogenous&quot;     
## [10] &quot;log10_total_features_by_counts_feature_control&quot;
## [11] &quot;log10_total_features_by_counts_Mt&quot;             
## [12] &quot;log10_total_features_by_counts_Ri&quot;             
## [13] &quot;log10_total_features_endogenous&quot;               
## [14] &quot;log10_total_features_feature_control&quot;          
## [15] &quot;log10_total_features_Mt&quot;                       
## [16] &quot;log10_total_features_Ri&quot;                       
## [17] &quot;pct_counts_endogenous&quot;                         
## [18] &quot;pct_counts_feature_control&quot;                    
## [19] &quot;pct_counts_in_top_100_features&quot;                
## [20] &quot;pct_counts_in_top_100_features_endogenous&quot;     
## [21] &quot;pct_counts_in_top_100_features_feature_control&quot;
## [22] &quot;pct_counts_in_top_100_features_Ri&quot;             
## [23] &quot;pct_counts_in_top_200_features&quot;                
## [24] &quot;pct_counts_in_top_200_features_endogenous&quot;     
## [25] &quot;pct_counts_in_top_50_features&quot;                 
## [26] &quot;pct_counts_in_top_50_features_endogenous&quot;      
## [27] &quot;pct_counts_in_top_50_features_feature_control&quot; 
## [28] &quot;pct_counts_in_top_50_features_Ri&quot;              
## [29] &quot;pct_counts_in_top_500_features&quot;                
## [30] &quot;pct_counts_in_top_500_features_endogenous&quot;     
## [31] &quot;pct_counts_Mt&quot;                                 
## [32] &quot;pct_counts_Ri&quot;                                 
## [33] &quot;pct_counts_top_100_features&quot;                   
## [34] &quot;pct_counts_top_100_features_endogenous&quot;        
## [35] &quot;pct_counts_top_100_features_feature_control&quot;   
## [36] &quot;pct_counts_top_100_features_Ri&quot;                
## [37] &quot;pct_counts_top_200_features&quot;                   
## [38] &quot;pct_counts_top_200_features_endogenous&quot;        
## [39] &quot;pct_counts_top_50_features&quot;                    
## [40] &quot;pct_counts_top_50_features_endogenous&quot;         
## [41] &quot;pct_counts_top_50_features_feature_control&quot;    
## [42] &quot;pct_counts_top_50_features_Ri&quot;                 
## [43] &quot;pct_counts_top_500_features&quot;                   
## [44] &quot;pct_counts_top_500_features_endogenous&quot;        
## [45] &quot;total_counts&quot;                                  
## [46] &quot;total_counts_endogenous&quot;                       
## [47] &quot;total_counts_feature_control&quot;                  
## [48] &quot;total_counts_Mt&quot;                               
## [49] &quot;total_counts_Ri&quot;                               
## [50] &quot;total_features&quot;                                
## [51] &quot;total_features_by_counts&quot;                      
## [52] &quot;total_features_by_counts_endogenous&quot;           
## [53] &quot;total_features_by_counts_feature_control&quot;      
## [54] &quot;total_features_by_counts_Mt&quot;                   
## [55] &quot;total_features_by_counts_Ri&quot;                   
## [56] &quot;total_features_endogenous&quot;                     
## [57] &quot;total_features_feature_control&quot;                
## [58] &quot;total_features_Mt&quot;                             
## [59] &quot;total_features_Ri&quot;
</code></pre>

<pre><code class="r">par(mfrow=c(2,2), mar=c(5, 4, 1, 1), bty=&quot;n&quot;)
hist(log10(sce$total_counts), xlab=&quot;log10(Library sizes)&quot;, main=&quot;&quot;, 
    breaks=20, col=&quot;grey80&quot;, ylab=&quot;Number of cells&quot;)

hist(log10(sce$total_features), xlab=&quot;log10(# of expressed genes)&quot;, 
     main=&quot;&quot;, breaks=20, col=&quot;grey80&quot;, ylab=&quot;Number of cells&quot;)

hist(sce$pct_counts_Ri, xlab=&quot;Ribosome prop. (%)&quot;,
    ylab=&quot;Number of cells&quot;, breaks=40, main=&quot;&quot;, col=&quot;grey80&quot;)

hist(sce$pct_counts_Mt, xlab=&quot;Mitochondrial prop. (%)&quot;, 
    ylab=&quot;Number of cells&quot;, breaks=80, main=&quot;&quot;, col=&quot;grey80&quot;)
</code></pre>

<p><img src="figure/unnamed-chunk-7-1.png" alt="plot of chunk unnamed-chunk-7"></p>

<pre><code class="r">par(mfrow=c(2,2), mar=c(5, 4, 1, 1), bty=&quot;n&quot;)
smoothScatter(log10(sce$total_counts), log10(sce$total_features), 
     xlab=&quot;log10(Library sizes)&quot;, ylab=&quot;log10(# of expressed genes)&quot;, 
     nrpoints=500, cex=0.5)
smoothScatter(log10(sce$total_counts), sce$pct_counts_Ri, 
     xlab=&quot;log10(Library sizes)&quot;, ylab=&quot;Ribosome prop. (%)&quot;,
     nrpoints=500, cex=0.5)
# abline(h=10,  lty=1)

smoothScatter(log10(sce$total_counts), sce$pct_counts_Mt, 
     xlab=&quot;log10(Library sizes)&quot;, ylab=&quot;Mitochondrial prop. (%)&quot;,
     nrpoints=500, cex=0.5)
# abline(h=5,  lty=1)

smoothScatter(sce$pct_counts_Ri, sce$pct_counts_Mt, 
     xlab=&quot;Ribosome prop. (%)&quot;, ylab=&quot;Mitochondrial prop. (%)&quot;,
     nrpoints=500, cex=0.5)
</code></pre>

<p><img src="figure/unnamed-chunk-7-2.png" alt="plot of chunk unnamed-chunk-7"></p>

<pre><code class="r"># abline(h=5,  lty=1)
# abline(v=10, lty=1)

par(mfrow=c(2,1))
hist(sce$pct_counts_Ri,breaks=40,col=&quot;gray&quot;)
hist(sce$pct_counts_Mt,breaks=40,col=&quot;gray&quot;)
</code></pre>

<p><img src="figure/unnamed-chunk-7-3.png" alt="plot of chunk unnamed-chunk-7"></p>

<pre><code class="r">par(mfrow=c(1,1))

plot(x=sce$pct_counts_Ri,y=sce$pct_counts_Mt,pch=16,col=rgb(0,0,0,0.4))
</code></pre>

<p><img src="figure/unnamed-chunk-7-4.png" alt="plot of chunk unnamed-chunk-7"></p>

<pre><code class="r"># abline(v=10,lty=2); abline(h=5,lty=2)

table(sce$pct_counts_Mt &lt; 5, sce$pct_counts_Ri &lt; 10)
</code></pre>

<pre><code>##        
##         FALSE  TRUE
##   FALSE     0    88
##   TRUE    171 14704
</code></pre>

<pre><code class="r">sce = sce[,which(sce$pct_counts_Mt &lt; 5 | sce$pct_counts_Ri &lt; 10)]
# sce = sce[,which(sce$pct_counts_Mt &lt; 5 &amp; sce$pct_counts_Ri &lt; 10)]
dim(sce)
</code></pre>

<pre><code>## [1] 31954 14963
</code></pre>

<h2>Summarize gene level information</h2>

<pre><code class="r">rowData(sce)[1:2,]
</code></pre>

<pre><code>## DataFrame with 2 rows and 20 columns
##   hgnc_symbol external_gene_name chromosome_name start_position
##   &lt;character&gt;        &lt;character&gt;     &lt;character&gt;      &lt;integer&gt;
## 1                    AC006946.16              22       17548711
## 2                    AC006946.17              22       17561591
##   end_position    strand description percentage_gene_gc_content
##      &lt;integer&gt; &lt;integer&gt; &lt;character&gt;                  &lt;numeric&gt;
## 1     17551565         1                                  41.61
## 2     17562346         1                                  41.67
##   gene_biotype is_feature_control is_feature_control_Mt
##    &lt;character&gt;          &lt;logical&gt;             &lt;logical&gt;
## 1      lincRNA              FALSE                 FALSE
## 2      lincRNA              FALSE                 FALSE
##   is_feature_control_Ri         mean_counts   log10_mean_counts
##               &lt;logical&gt;           &lt;numeric&gt;           &lt;numeric&gt;
## 1                 FALSE 0.00548018445498897  0.0023735161394708
## 2                 FALSE 0.00360890195816347 0.00156450482886486
##   n_cells_by_counts pct_dropout_by_counts total_counts log10_total_counts
##           &lt;integer&gt;             &lt;numeric&gt;    &lt;integer&gt;          &lt;numeric&gt;
## 1                79      99.4720310098242           82   1.91907809237607
## 2                52      99.6524761077324           54   1.74036268949424
##   n_cells_counts pct_dropout_counts
##        &lt;integer&gt;          &lt;numeric&gt;
## 1             79   99.4720310098242
## 2             52   99.6524761077324
</code></pre>

<pre><code class="r">min(rowData(sce)$mean_counts)
</code></pre>

<pre><code>## [1] 6.683152e-05
</code></pre>

<pre><code class="r">min(rowData(sce)$mean_counts[rowData(sce)$mean_counts&gt;0])
</code></pre>

<pre><code>## [1] 6.683152e-05
</code></pre>

<pre><code class="r">min(rowData(sce)$n_cells_counts)
</code></pre>

<pre><code>## [1] 1
</code></pre>

<pre><code class="r">par(mfrow=c(1,3), mar=c(5,4,1,1))
hist(log10(rowData(sce)$mean_counts+1e-6), col=&quot;grey80&quot;,  main=&quot;&quot;, 
     breaks=40, xlab=&quot;log10(ave # of UMI + 1e-6)&quot;)
hist(log10(rowData(sce)$n_cells_counts+1), col=&quot;grey80&quot;, main=&quot;&quot;, 
     breaks=40, xlab=&quot;log10(# of expressed cells + 1)&quot;)
smoothScatter(log10(rowData(sce)$mean_counts+1e-6), 
              log10(rowData(sce)$n_cells_counts + 1), 
              xlab=&quot;log10(ave # of UMI + 1e-6)&quot;, 
              ylab=&quot;log10(# of expressed cells + 1)&quot;)
</code></pre>

<p><img src="figure/unnamed-chunk-8-1.png" alt="plot of chunk unnamed-chunk-8"></p>

<pre><code class="r">par(mfrow=c(1,1))

tb1 = table(rowData(sce)$n_cells_counts)
tb1[1:11]
</code></pre>

<pre><code>## 
##    1    2    3    4    5    6    7    8    9   10   11 
## 3308 1960 1281 1031  843  667  552  526  473  383  369
</code></pre>

<p>We remove those genes that are expressed in zero or only one cell. The variable <em>strand</em> need to be renamed, otherwise there is an error message that such a variabel name cannot be used. </p>

<pre><code class="r">names(rowData(sce))[6] = &quot;strand_n&quot;

# Dr. Sun&#39;s code
rowData(sce)$n_cell_count_filter = rowData(sce)$n_cells_counts &gt; 1
table(rowData(sce)$n_cell_count_filter)
</code></pre>

<pre><code>## 
## FALSE  TRUE 
##  3308 28646
</code></pre>

<pre><code class="r"># Based on paper&#39;s methods: Filter for genes
rowData(sce)$gene_detect_filter = apply(counts(sce),1,function(xx) length(xx[xx &gt;= 2]) &gt;= 10)
# Aka a gene in a cell is considered &quot;detected&quot; if it has at least 2 UMIs(counts),
# Retain a gene if it has been detected in at least 10 cells(samples)
table(rowData(sce)$gene_detect_filter)
</code></pre>

<pre><code>## 
## FALSE  TRUE 
## 20669 11285
</code></pre>

<pre><code class="r"># Based on paper&#39;s methods: Filter for nuclei
colData(sce)$sample_filter = apply(counts(sce),2,function(xx) length(xx[xx &gt;= 2]) &gt;= 200)
# Same definition for a gene in a cell detected
# Retain a sample if it has at least 200 genes detected
table(colData(sce)$sample_filter)
</code></pre>

<pre><code>## 
## FALSE  TRUE 
## 12899  2064
</code></pre>

<pre><code class="r"># sce = sce[which(rowData(sce)$n_cell_count_filter == TRUE),] # less stringent gene filter
# sce = sce[which(rowData(sce)$gene_detect_filter == TRUE),] # more stringent gene filter

# No gene/sample filter applied, I&#39;m assuming the paper did analysis with all available genes and samples provided in the online dataset!!
</code></pre>

<p>Next we check those highly expressed genes </p>

<pre><code class="r">par(mar=c(5,4,1,1))
od1 = order(rowData(sce)$mean_counts, decreasing = TRUE)
barplot(rowData(sce)$mean_counts[od1[20:1]], las=1, 
        names.arg=rowData(sce)$hgnc_symbol[od1[20:1]], 
        horiz=TRUE, cex.names=0.8, cex.axis=0.8, 
        xlab=&quot;ave # of UMI&quot;)
</code></pre>

<p><img src="figure/unnamed-chunk-10-1.png" alt="plot of chunk unnamed-chunk-10"></p>

<h2>Normalization</h2>

<p>A simple solution for normalization and stablizing expression varaince across genes is to tranform the count data by log(count/size.factor + 1). One may calcualte size.factor per cell as the total number of UMIs, and this assumes the total expression are the same across all the cells. However, the total expression of each cell may vary with respect to cell type and/or cell size, and the <code>computeSumFactors</code> function in R package scran provides a more  sophisicated way to calcualte size.factor to allow such variaation across cells [@lun2016pooling]. <code>computeSumFactors</code> can use initial clustering of cells to normalize expression within and beetween clusters.  Within a cluster, it estimates the size factor for many groups of cells so that there are more groups than cells, and then it can calcualte the size factor per cell using a lienar deconvolution system. </p>

<p>As shown in the following plot, the final size factor estimation is indeed highly correlated with the naive definition by total count. </p>

<p>Finally, the command <code>normalize(sce)</code> adds the normalized expression into the variable <code>sce</code>.</p>

<pre><code class="r">date()
</code></pre>

<pre><code>## [1] &quot;Sat Sep 29 01:25:35 2018&quot;
</code></pre>

<pre><code class="r">clusters = quickCluster(sce, min.mean=0.1, method=&quot;igraph&quot;)
table(clusters)
</code></pre>

<pre><code>## clusters
##    1    2    3    4    5 
## 1370 2568 2029 3265 5731
</code></pre>

<pre><code class="r">date()
</code></pre>

<pre><code>## [1] &quot;Sat Sep 29 01:26:52 2018&quot;
</code></pre>

<pre><code class="r">sce = computeSumFactors(sce, cluster=clusters, min.mean=0.1)
date()
</code></pre>

<pre><code>## [1] &quot;Sat Sep 29 01:30:52 2018&quot;
</code></pre>

<pre><code class="r">summary(sizeFactors(sce))
</code></pre>

<pre><code>##      Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
##  0.003812  0.422892  0.741574  1.000000  1.248273 12.069274
</code></pre>

<pre><code class="r"># hist(sizeFactors(sce),breaks=50,col=&quot;gray&quot;)

# Remove cells with negative or very small size factors
dim(sce)
</code></pre>

<pre><code>## [1] 31954 14963
</code></pre>

<pre><code class="r">sce = sce[,which(sizeFactors(sce) &gt; 0)]
dim(sce)
</code></pre>

<pre><code>## [1] 31954 14963
</code></pre>

<pre><code class="r">par(mfrow=c(1,2), mar=c(5,4,2,1), bty=&quot;n&quot;)
smoothScatter(sce$total_counts, sizeFactors(sce), log=&quot;xy&quot;, 
              xlab=&quot;total counts&quot;, ylab=&quot;size factors&quot;)
plot(sce$total_counts, sizeFactors(sce), log=&quot;xy&quot;, 
     xlab=&quot;total counts&quot;, ylab=&quot;size factors&quot;, 
     cex=0.3, pch=20, col=rgb(0.1,0.2,0.7,0.3))
</code></pre>

<p><img src="figure/unnamed-chunk-11-1.png" alt="plot of chunk unnamed-chunk-11"></p>

<pre><code class="r">sce = normalize(sce) 
# logcounts(sce) = log2(gene_cell_count / size_factor + 1)
# now the cells are comparable with each other for dimension reduction/clustering

# Save temp image
post_norm_fn = file.path(data_dir,&quot;post_norm.rds&quot;)
if( where_file == &quot;longleaf&quot; ){
  saveRDS(sce,post_norm_fn)
} else if(FALSE){
  sce = readRDS(post_norm_fn)
}
</code></pre>

<h2>Dimension Reduction</h2>

<p>For dimension reduction, such as calculating PCA or performing TSNE, we should start by identifying a subset of genes with high level of biological signal relative to background (technical) noise. The <code>decomposeVar</code> function from R/cran is designed for this task. </p>

<pre><code class="r">new.trend = makeTechTrend(x=sce)
fit = trendVar(sce, use.spikes=FALSE, loess.args=list(span=0.05))

par(mfrow=c(1,1), mar=c(5,4,2,1), bty=&quot;n&quot;)
plot(fit$mean, fit$var, pch=20, col=rgb(0.1,0.2,0.7,0.6), 
     xlab=&quot;log(mean)&quot;, ylab=&quot;var&quot;)
curve(fit$trend(x), col=&quot;orange&quot;, lwd=2, add=TRUE)
curve(new.trend(x), col=&quot;red&quot;, lwd=2, add=TRUE)
legend(&quot;top&quot;, legend=c(&quot;Poisson noise&quot;, &quot;observed trend&quot;), 
       lty=1, lwd=2, col=c(&quot;red&quot;, &quot;orange&quot;), bty=&quot;n&quot;)
</code></pre>

<p><img src="figure/unnamed-chunk-12-1.png" alt="plot of chunk unnamed-chunk-12"></p>

<pre><code class="r">fit$trend = new.trend
dec = decomposeVar(fit=fit) # obtain bio, FDR, pvalues for testing for HVGs (highly variable genes)
top.dec = dec[order(dec$bio, decreasing=TRUE),]
plotExpression(sce, features=rownames(top.dec)[1:10])
</code></pre>

<p><img src="figure/unnamed-chunk-12-2.png" alt="plot of chunk unnamed-chunk-12"></p>

<p>When performing PCA, we can use all the genes or just those genes with high signal-to-noise ratio. TSNE analysis is usually based on the top PCs rather than the original gene expression data. We first perform PCA using all the genes and the function <code>denoisePCA</code> can automatically select the PCs based on modeling of technical noise. </p>

<pre><code class="r">date()
</code></pre>

<pre><code>## [1] &quot;Sat Sep 29 01:33:31 2018&quot;
</code></pre>

<pre><code class="r"># This will basically objectively select the appropriate number of PCs
sce = denoisePCA(sce, technical=new.trend, approx=TRUE)
date()
</code></pre>

<pre><code>## [1] &quot;Sat Sep 29 01:40:40 2018&quot;
</code></pre>

<pre><code class="r">dim(reducedDim(sce, &quot;PCA&quot;))
</code></pre>

<pre><code>## [1] 14963   100
</code></pre>

<pre><code class="r">plot(log10(attr(reducedDim(sce), &quot;percentVar&quot;)), xlab=&quot;PC&quot;,
     ylab=&quot;log10(Prop of variance explained)&quot;, pch=20, cex=0.6, 
     col=rgb(0.8, 0.2, 0.2, 0.5))
abline(v=ncol(reducedDim(sce, &quot;PCA&quot;)), lty=2, col=&quot;red&quot;)
</code></pre>

<p><img src="figure/unnamed-chunk-14-1.png" alt="plot of chunk unnamed-chunk-14"></p>

<pre><code class="r">df_pcs = data.frame(reducedDim(sce, &quot;PCA&quot;))
df_pcs$log10_total_features = colData(sce)$log10_total_features
df_pcs$part_cell_id = sapply(colnames(sce),function(xx) 
  strsplit(xx,&quot;_&quot;)[[1]][1],USE.NAMES=FALSE)

gp1 = ggplot(df_pcs, aes(PC1,PC2,col=log10_total_features)) + 
  geom_point(size=0.2,alpha=0.6) + theme_classic() + 
  scale_colour_gradient(low=&quot;lightblue&quot;,high=&quot;red&quot;) +
  guides(color = guide_legend(override.aes = list(size=3)))
gp1
</code></pre>

<p><img src="figure/unnamed-chunk-14-2.png" alt="plot of chunk unnamed-chunk-14"></p>

<pre><code class="r">range(df_pcs$PC1)
</code></pre>

<pre><code>## [1]  -5.179616 145.941758
</code></pre>

<pre><code class="r">range(df_pcs$PC2)
</code></pre>

<pre><code>## [1] -35.56816  10.77493
</code></pre>

<pre><code class="r">table(df_pcs$part_cell_id)
</code></pre>

<pre><code>## 
##          hCc          hCd          hCe          hCf         hHP1 
##          780          812          733          779          486 
##         hHP2        hHP2a        hHP2b        hHP2c        hHP3b 
##          320          144          166          115           47 
##        HP2-A        HP2-B        HP3-A        HP3-B HP3-B-united 
##          717          674         1012         1014         1232 
##    humanPFCa    humanPFCb       PFC-CD      PFC2-A1      PFC2-A2 
##          752          659          918         1092          953 
##      PFC2-A3      PFC2-A5 
##         1085          473
</code></pre>

<pre><code class="r">table(df_pcs$part_cell_id[which(df_pcs$PC1 &gt;= 25)])
</code></pre>

<pre><code>## 
##  hHP1  hHP2 hHP2a hHP2b hHP2c hHP3b 
##    37    34    27    25    19     7
</code></pre>

<pre><code class="r">gp2 = ggplot(df_pcs, aes(PC1,PC2,col=part_cell_id)) + 
  geom_point(size=0.5,alpha=0.6) + theme_classic() +
  guides(color = guide_legend(override.aes = list(size=3)))
gp2
</code></pre>

<p><img src="figure/unnamed-chunk-14-3.png" alt="plot of chunk unnamed-chunk-14"></p>

<pre><code class="r">date()
</code></pre>

<pre><code>## [1] &quot;Sat Sep 29 01:40:42 2018&quot;
</code></pre>

<pre><code class="r">sce = runTSNE(sce, use_dimred=&quot;PCA&quot;, perplexity=30, rand_seed=100)
date()
</code></pre>

<pre><code>## [1] &quot;Sat Sep 29 01:46:43 2018&quot;
</code></pre>

<pre><code class="r">df_tsne = data.frame(reducedDim(sce, &quot;TSNE&quot;))
df_tsne$log10_total_features = colData(sce)$log10_total_features
df_tsne$part_cell_id = sapply(rownames(df_pcs),function(xx) 
  strsplit(xx,&quot;_&quot;)[[1]][1],USE.NAMES=FALSE)

gp1 = ggplot(df_tsne, aes(X1,X2,col=log10_total_features)) + 
  geom_point(size=0.2,alpha=0.6) + theme_classic() + 
  scale_colour_gradient(low=&quot;lightblue&quot;,high=&quot;red&quot;) +
  guides(color = guide_legend(override.aes = list(size=3)))
gp1
</code></pre>

<p><img src="figure/unnamed-chunk-14-4.png" alt="plot of chunk unnamed-chunk-14"></p>

<pre><code class="r">gp2 = ggplot(df_tsne, aes(X1,X2,col=part_cell_id)) + 
  geom_point(size=0.5,alpha=0.6) + theme_classic() + 
  # scale_colour_gradient(low=&quot;lightblue&quot;,high=&quot;red&quot;) +
  guides(color = guide_legend(override.aes = list(size=3)))
gp2
</code></pre>

<p><img src="figure/unnamed-chunk-14-5.png" alt="plot of chunk unnamed-chunk-14"></p>

<pre><code class="r"># Save temp image
post_dimRed_fn = file.path(data_dir,&quot;post_dimRed.rds&quot;)
if( where_file == &quot;longleaf&quot; ){
  saveRDS(list(sce=sce,dec=dec),post_dimRed_fn)
} else if(FALSE){
  rds = readRDS(post_dimRed_fn)
  sce = rds$sce
  dec = rds$dec
  rm(rds)
}
</code></pre>

<p>Next we only select around top 1000 genes for the PCA and use the top 50 PCs for TSNE projection. </p>

<pre><code class="r">library(svd)
library(Rtsne)

summary(dec$bio)
</code></pre>

<pre><code>##      Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
## -0.021880 -0.000035  0.000148  0.004161  0.002228  5.124444
</code></pre>

<pre><code class="r">dec1 = dec
dec1$bio[which(dec$bio &lt; 1e-8)] = 1e-8
dec1$FDR[which(dec$FDR &lt; 1e-100)] = 1e-100

par(mfrow=c(1,2))
hist(log10(dec1$bio), breaks=100, main=&quot;&quot;)
hist(log10(dec1$FDR), breaks=100, main=&quot;&quot;)
</code></pre>

<p><img src="figure/unnamed-chunk-15-1.png" alt="plot of chunk unnamed-chunk-15"></p>

<pre><code class="r">summary(dec$FDR[dec$bio &gt; 0.001])
</code></pre>

<pre><code>##     Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
## 0.000000 0.000000 0.000000 0.002185 0.000000 0.559689
</code></pre>

<pre><code class="r">table(dec$FDR &lt; 1e-10, dec$bio &gt; 0.01)
</code></pre>

<pre><code>##        
##         FALSE  TRUE
##   FALSE 16170    84
##   TRUE  12944  2756
</code></pre>

<pre><code class="r"># Subsetting genes based on FDR and biological residual thresholds
w2kp = which(dec$FDR &lt; 1e-10 &amp; dec$bio &gt; 0.01) # original gene filter
sce_sub = sce[w2kp,]
sce_sub
</code></pre>

<pre><code>## class: SingleCellExperiment 
## dim: 2756 14963 
## metadata(1): log.exprs.offset
## assays(2): counts logcounts
## rownames(2756): AATK AC004158.3 ... ZSWIM6 ZNF704
## rowData names(22): hgnc_symbol external_gene_name ...
##   n_cell_count_filter gene_detect_filter
## colnames(14963): hHP1_AACACTATCTAC hHP1_CTACGCATCCAT ...
##   PFC-CD_CTCCATTCATGC PFC-CD_CGTCATTAGCAG
## colData names(60): is_cell_control total_features_by_counts ...
##   pct_counts_top_100_features_Ri sample_filter
## reducedDimNames(2): PCA TSNE
## spikeNames(0):
</code></pre>

<pre><code class="r"># Extracting log2(norm_express+1), transposing, scaling columns(genes) to have mean = 0, var = 1
edat = t(as.matrix(logcounts(sce_sub)))
edat = scale(edat)
dim(edat)
</code></pre>

<pre><code>## [1] 14963  2756
</code></pre>

<pre><code class="r">edat[1:2,1:3]
</code></pre>

<pre><code>##                         AATK AC004158.3      ABCF1
## hHP1_AACACTATCTAC -0.3204690 -0.1782094 -0.2398015
## hHP1_CTACGCATCCAT  0.2698021 -0.1782094  1.3959415
</code></pre>

<pre><code class="r"># Perform SVD on sce_sub
date()
</code></pre>

<pre><code>## [1] &quot;Sat Sep 29 01:48:22 2018&quot;
</code></pre>

<pre><code class="r">ppk = propack.svd(edat,neig=50)
date()
</code></pre>

<pre><code>## [1] &quot;Sat Sep 29 01:49:13 2018&quot;
</code></pre>

<pre><code class="r">pca = t(ppk$d*t(ppk$u)) # calculates pc scores aka principal components

df_pcs = data.frame(pca)
df_pcs$log10_total_features = colData(sce_sub)$log10_total_features
df_pcs$part_cell_id = sapply(colnames(sce_sub),function(xx) 
  strsplit(xx,&quot;_&quot;)[[1]][1],USE.NAMES=FALSE)
df_pcs[1:2,1:10]
</code></pre>

<pre><code>##           X1        X2       X3         X4        X5        X6        X7
## 1 -0.5224780 -3.526958 2.688059 -0.4063681 -0.881605 -2.419041 1.2778257
## 2  0.3728504 -3.365366 1.632852  1.0890913 -1.639625 -2.626520 0.6192533
##           X8       X9        X10
## 1 -0.3760573 1.113652 -0.5036978
## 2 -1.0294712 2.835417 -0.5974646
</code></pre>

<pre><code class="r">gp1 = ggplot(df_pcs, aes(X1,X2,col=log10_total_features)) + 
  geom_point(size=0.2,alpha=0.6) + theme_classic() + 
  scale_colour_gradient(low=&quot;lightblue&quot;,high=&quot;red&quot;) +
  guides(color = guide_legend(override.aes = list(size=3)))
gp1
</code></pre>

<p><img src="figure/unnamed-chunk-15-2.png" alt="plot of chunk unnamed-chunk-15"></p>

<pre><code class="r">gp2 = ggplot(df_pcs, aes(X1,X2,col=part_cell_id)) + 
  geom_point(size=0.5,alpha=0.6) + theme_classic() + 
  # scale_colour_gradient(low=&quot;lightblue&quot;,high=&quot;red&quot;) +
  guides(color = guide_legend(override.aes = list(size=3)))
gp2
</code></pre>

<p><img src="figure/unnamed-chunk-15-3.png" alt="plot of chunk unnamed-chunk-15"></p>

<pre><code class="r">set.seed(100)
date()
</code></pre>

<pre><code>## [1] &quot;Sat Sep 29 01:49:15 2018&quot;
</code></pre>

<pre><code class="r">tsne = Rtsne(pca, pca = FALSE)
date()
</code></pre>

<pre><code>## [1] &quot;Sat Sep 29 01:54:05 2018&quot;
</code></pre>

<pre><code class="r">df_tsne = data.frame(tsne$Y)
df_tsne$log10_total_features = colData(sce_sub)$log10_total_features
df_tsne$part_cell_id = sapply(colnames(sce_sub),function(xx) 
  strsplit(xx,&quot;_&quot;)[[1]][1],USE.NAMES=FALSE)
dim(df_tsne)
</code></pre>

<pre><code>## [1] 14963     4
</code></pre>

<pre><code class="r">df_tsne[1:2,]
</code></pre>

<pre><code>##         X1       X2 log10_total_features part_cell_id
## 1 1.278823 19.99260             3.393224         hHP1
## 2 2.902783 18.11585             3.453777         hHP1
</code></pre>

<pre><code class="r">gp1 = ggplot(df_tsne, aes(X1,X2,col=log10_total_features)) + 
  geom_point(size=0.2,alpha=0.6) + theme_classic() + 
  scale_colour_gradient(low=&quot;lightblue&quot;,high=&quot;red&quot;) +
  guides(color = guide_legend(override.aes = list(size=3)))
gp1
</code></pre>

<p><img src="figure/unnamed-chunk-15-4.png" alt="plot of chunk unnamed-chunk-15"></p>

<pre><code class="r">gp2 = ggplot(df_tsne, aes(X1,X2,col=part_cell_id)) + 
  geom_point(size=0.2,alpha=0.6) + theme_classic() + 
  # scale_colour_gradient(low=&quot;lightblue&quot;,high=&quot;red&quot;) +
  guides(color = guide_legend(override.aes = list(size=3)))
gp2
</code></pre>

<p><img src="figure/unnamed-chunk-15-5.png" alt="plot of chunk unnamed-chunk-15"></p>

<pre><code class="r">reducedDims(sce_sub) = SimpleList(PCA=pca, TSNE=tsne$Y)
sce_sub
</code></pre>

<pre><code>## class: SingleCellExperiment 
## dim: 2756 14963 
## metadata(1): log.exprs.offset
## assays(2): counts logcounts
## rownames(2756): AATK AC004158.3 ... ZSWIM6 ZNF704
## rowData names(22): hgnc_symbol external_gene_name ...
##   n_cell_count_filter gene_detect_filter
## colnames(14963): hHP1_AACACTATCTAC hHP1_CTACGCATCCAT ...
##   PFC-CD_CTCCATTCATGC PFC-CD_CGTCATTAGCAG
## colData names(60): is_cell_control total_features_by_counts ...
##   pct_counts_top_100_features_Ri sample_filter
## reducedDimNames(2): PCA TSNE
## spikeNames(0):
</code></pre>

<h2>Clustering</h2>

<h3>Kmeans</h3>

<p>There are many methods for clustering of single cell RNA-seq data. The performance of each method may also depend on pre-processing steps, such as performing imputation or not. We wil compare these methods in a seperate document. Here we just illustrate the clustering reuslts using a simple kmeans method on the top 50 PCs. </p>

<pre><code class="r">k10_50_pcs = kmeans(reducedDim(sce_sub, &quot;PCA&quot;), 
                    # centers=10, # original choice 
                    centers = 15, # newer choice
                    iter.max=150, algorithm=&quot;MacQueen&quot;)
names(k10_50_pcs)
</code></pre>

<pre><code>## [1] &quot;cluster&quot;      &quot;centers&quot;      &quot;totss&quot;        &quot;withinss&quot;    
## [5] &quot;tot.withinss&quot; &quot;betweenss&quot;    &quot;size&quot;         &quot;iter&quot;        
## [9] &quot;ifault&quot;
</code></pre>

<pre><code class="r">dim(k10_50_pcs$centers)
</code></pre>

<pre><code>## [1] 15 50
</code></pre>

<pre><code class="r">df_tsne$cluster_kmean = as.factor(k10_50_pcs$cluster)
# cols = c(&quot;#FB9A99&quot;,&quot;#FF7F00&quot;,&quot;yellow&quot;,&quot;orchid&quot;,&quot;grey&quot;,&quot;red&quot;,&quot;dodgerblue2&quot;,&quot;tan4&quot;,&quot;green4&quot;,&quot;#99c9fb&quot;)

gp1 = ggplot(df_tsne, aes(X1,X2,col=cluster_kmean)) + 
  geom_point(size=0.2,alpha=0.6) + theme_classic() + 
  # scale_color_manual(values=cols) + 
  guides(color = guide_legend(override.aes = list(size=3)))
gp1
</code></pre>

<p><img src="figure/unnamed-chunk-16-1.png" alt="plot of chunk unnamed-chunk-16"></p>

<pre><code class="r"># Save temp image
post_cluster_fn = file.path(data_dir,&quot;post_cluster.rds&quot;)
if(where_file == &quot;longleaf&quot;){
  saveRDS(list(sce=sce,sce_sub=sce_sub,
    k10_50_pcs=k10_50_pcs,df_tsne=df_tsne),
    post_cluster_fn)
} else if(FALSE){
  rds = readRDS(post_cluster_fn)
  sce = rds$sce
  sce_sub = rds$sce_sub
  k10_50_pcs = rds$k10_50_pcs
  df_tsne = rds$df_tsne
  rm(rds)
}
</code></pre>

<p>So in the paper, they have 15 clusters (ASC1,ASC2,END,exCA1,exCA3,exDG,exPFC1,exPFC2,GABA1,GABA2,MG,NSC,ODC1,OPC, and Unclassified). Using Dr. Sun&#39;s code, we specified 15 clusters for kmeans by clustering 50 PCAs derived from filtering down to a set of 2756 genes. </p>

<h3>SC3</h3>

<p>Code used here is based on <a href="https://bioconductor.org/packages/devel/bioc/vignettes/SC3/inst/doc/SC3.html#run-sc3">this link</a>.</p>

<pre><code class="r">library(SC3)
rowData(sce)$feature_symbol = rowData(sce)$external_gene_name
time1 = date()
# all_ks = c(2,4)
all_ks = c(10,15)
sce = sc3(sce, 
          # ks = c(2,4),
          ks = all_ks,biology = TRUE,
          n_cores = 1,rand_seed = 100,
          svm_num_cells = 1000
          # svm_num_cells = 5000
          )
</code></pre>

<pre><code>## Setting SC3 parameters...
</code></pre>

<pre><code>## Your dataset contains more than 2000 cells. Adjusting the nstart parameter of kmeans to 50 for faster performance...
</code></pre>

<pre><code>## Defining training cells for SVM using svm_num_cells parameter...
</code></pre>

<pre><code>## Calculating distances between the cells...
</code></pre>

<pre><code>## Performing transformations and calculating eigenvectors...
</code></pre>

<pre><code>## Performing k-means clustering...
</code></pre>

<pre><code>## Calculating consensus matrix...
</code></pre>

<pre><code>## Calculating biology...
</code></pre>

<pre><code class="r">warnings()
time2 = date()
time1; time2
</code></pre>

<pre><code>## [1] &quot;Sat Sep 29 01:55:50 2018&quot;
</code></pre>

<pre><code>## [1] &quot;Sat Sep 29 02:46:49 2018&quot;
</code></pre>

<pre><code class="r"># Run SVM and predict labels of all other cells
time3 = date()
sce = sc3_run_svm(sce, ks = all_ks)
time4 = date()
time3; time4
</code></pre>

<pre><code>## [1] &quot;Sat Sep 29 02:46:49 2018&quot;
</code></pre>

<pre><code>## [1] &quot;Sat Sep 29 02:48:53 2018&quot;
</code></pre>

<pre><code class="r"># Save temp image
post_sc3_fn = file.path(data_dir,&quot;post_sc3.rds&quot;)
if(where_file == &quot;longleaf&quot;){
  saveRDS(list(sce=sce,all_ks=all_ks,k10_50_pcs=k10_50_pcs,df_tsne=df_tsne,sce_sub=sce_sub),post_sc3_fn)
} else if(FALSE){
  rds = readRDS(post_sc3_fn)
  sce = rds$sce
  all_ks = rds$all_ks
  df_tsne = rds$df_tsne
  rm(rds)
}

# Plotting
for(one_ks in all_ks){
  # one_ks = 2
  print(one_ks)
  plotPCA(sce,
      colour_by = paste0(&quot;sc3_&quot;,one_ks,&quot;_clusters&quot;), 
      size_by = paste0(&quot;sc3_&quot;,one_ks,&quot;_log2_outlier_score&quot;))

  sc3_plot_consensus(sce,k = one_ks,
      show_pdata = c(&quot;log10_total_features&quot;,
        paste0(&quot;sc3_&quot;,one_ks,&quot;_clusters&quot;), 
        paste0(&quot;sc3_&quot;,one_ks,&quot;_log2_outlier_score&quot;)))

  sc3_plot_silhouette(sce, k = one_ks)

  sc3_plot_expression(sce, k = one_ks,
      show_pdata = c(&quot;log10_total_features&quot;,
        paste0(&quot;sc3_&quot;,one_ks,&quot;_clusters&quot;), 
        paste0(&quot;sc3_&quot;,one_ks,&quot;_log2_outlier_score&quot;)))

  sc3_plot_cluster_stability(sce, k = one_ks)

  sc3_plot_de_genes(sce, k = one_ks, 
    show_pdata = c(&quot;log10_total_features&quot;,
        paste0(&quot;sc3_&quot;,one_ks,&quot;_clusters&quot;), 
        paste0(&quot;sc3_&quot;,one_ks,&quot;_log2_outlier_score&quot;)))

  sc3_plot_markers(sce, k = one_ks, 
    show_pdata = c(&quot;log10_total_features&quot;,
        paste0(&quot;sc3_&quot;,one_ks,&quot;_clusters&quot;), 
        paste0(&quot;sc3_&quot;,one_ks,&quot;_log2_outlier_score&quot;)))

}
</code></pre>

<pre><code>## [1] 10
</code></pre>

<p><img src="figure/unnamed-chunk-17-1.png" alt="plot of chunk unnamed-chunk-17"><img src="figure/unnamed-chunk-17-2.png" alt="plot of chunk unnamed-chunk-17"><img src="figure/unnamed-chunk-17-3.png" alt="plot of chunk unnamed-chunk-17"><img src="figure/unnamed-chunk-17-4.png" alt="plot of chunk unnamed-chunk-17"><img src="figure/unnamed-chunk-17-5.png" alt="plot of chunk unnamed-chunk-17"></p>

<pre><code>## [1] 15
</code></pre>

<p><img src="figure/unnamed-chunk-17-6.png" alt="plot of chunk unnamed-chunk-17"><img src="figure/unnamed-chunk-17-7.png" alt="plot of chunk unnamed-chunk-17"><img src="figure/unnamed-chunk-17-8.png" alt="plot of chunk unnamed-chunk-17"><img src="figure/unnamed-chunk-17-9.png" alt="plot of chunk unnamed-chunk-17"><img src="figure/unnamed-chunk-17-10.png" alt="plot of chunk unnamed-chunk-17"></p>

<h2>Compare our kmeans to SC3 and to DrocNc clustering</h2>

<p>Compare paper&#39;s clustering to ours: check if paper supplement cluster is consistent with our results, check how many genes the paper have left</p>

<pre><code class="r">source(file.path(src_dir,&quot;SOURCE.R&quot;))

tmp_lab = smart_RT(file.path(data_dir,&quot;cluster_num_label.txt&quot;),sep = &quot;\t&quot;,header = TRUE)
tmp_lab = name_change(tmp_lab,&quot;Name&quot;,&quot;Cluster.Name&quot;)
tmp_lab = name_change(tmp_lab,&quot;Name.1&quot;,&quot;Cell_Type&quot;)

tmp_res = smart_RT(file.path(data_dir,&quot;paper_cluster_res.txt&quot;),sep = &quot;\t&quot;,header = TRUE,comment.char = &quot;&quot;)
tmp_res = smart_merge(tmp_res,tmp_lab[,c(&quot;Cluster.Name&quot;,&quot;Cell_Type&quot;)],all.x=TRUE)

table(tmp_res[,c(&quot;Cluster.ID&quot;,&quot;Cluster.Name&quot;)])
</code></pre>

<pre><code>##           Cluster.Name
## Cluster.ID ASC1 ASC2  END exCA1 exCA3 exDG exPFC1 exPFC2 GABA1 GABA2   MG
##         1     0    0    0     0     0    0   3104      0     0     0    0
##         2     0    0    0     0     0    0      0    297     0     0    0
##         3     0    0    0   421     0    0      0      0     0     0    0
##         4     0    0    0     0   745    0      0      0     0     0    0
##         5     0    0    0     0     0    0      0      0   892     0    0
##         6     0    0    0     0     0    0      0      0     0   823    0
##         7     0    0    0     0     0 1453      0      0     0     0    0
##         8  1204    0    0     0     0    0      0      0     0     0    0
##         9     0  705    0     0     0    0      0      0     0     0    0
##         10    0    0    0     0     0    0      0      0     0     0    0
##         11    0    0    0     0     0    0      0      0     0     0    0
##         12    0    0    0     0     0    0      0      0     0     0    0
##         13    0    0    0     0     0    0      0      0     0     0  389
##         14    0    0    0     0     0    0      0      0     0     0    0
##         15    0    0  254     0     0    0      0      0     0     0    0
##         16    0    0    0     0     0    0      0      0     0     0    0
##         17    0    0    0     0     0    0      0      0     0     0    0
##         18    0    0    0     0     0    0      0      0     0     0    0
##           Cluster.Name
## Cluster.ID  NSC ODC1  OPC Unclassified
##         1     0    0    0            0
##         2     0    0    0            0
##         3     0    0    0            0
##         4     0    0    0            0
##         5     0    0    0            0
##         6     0    0    0            0
##         7     0    0    0            0
##         8     0    0    0            0
##         9     0    0    0            0
##         10    0 1718    0            0
##         11    0 1247    0            0
##         12    0    0  684            0
##         13    0    0    0            0
##         14  201    0    0            0
##         15    0    0    0            0
##         16    0    0    0          200
##         17    0    0    0          132
##         18    0    0    0          494
</code></pre>

<pre><code class="r">range(tmp_res$X.Genes)
</code></pre>

<pre><code>## [1]  201 4249
</code></pre>

<pre><code class="r">hist(tmp_res$X.Genes,breaks=50,col=&quot;gray&quot;)
</code></pre>

<p><img src="figure/unnamed-chunk-18-1.png" alt="plot of chunk unnamed-chunk-18"></p>

<pre><code class="r">df_tsne$Cell.ID = colnames(sce_sub)

# Merge and compare
all_clust_res = smart_merge(df_tsne,tmp_res)
sc3_res = smart_df(colData(sce)[,c(&quot;sc3_10_clusters&quot;,&quot;sc3_15_clusters&quot;)])
sc3_res$Cell.ID = colnames(sce)
all_clust_res = smart_merge(all_clust_res,sc3_res)
all_clust_res[1:5,]
</code></pre>

<pre><code>##            Cell.ID        X1         X2 log10_total_features part_cell_id
## 1 hCc_AAAAAGCTACAA -4.505137   3.956237             2.866878          hCc
## 2 hCc_AAACAGGTGAGG 15.932901 -10.190275             3.170262          hCc
## 3 hCc_AAACCCTTTACA 13.252554 -10.390478             3.044540          hCc
## 4 hCc_AAACGTGACGGA 14.660336 -11.524911             2.849419          hCc
## 5 hCc_AAAGAACTCGCG -4.715978   1.991771             2.716838          hCc
##   cluster_kmean Cluster.Name X.Genes X.Transcripts Cluster.ID Cell_Type
## 1            14       exPFC1     736          1034          1     exPFC
## 2             2        GABA1    1480          2391          5      GABA
## 3             5        GABA1    1107          1757          5      GABA
## 4             5        GABA1     708           935          5      GABA
## 5            14       exPFC1     520           673          1     exPFC
##   sc3_10_clusters sc3_15_clusters
## 1               5               1
## 2               3              10
## 3               3              10
## 4               3              10
## 5               5               1
</code></pre>

<pre><code class="r">smart_table(all_clust_res[,c(&quot;Cell_Type&quot;,&quot;Cluster.Name&quot;)])
</code></pre>

<pre><code>##          Cluster.Name
## Cell_Type ASC1 ASC2  END exCA1 exCA3 exDG exPFC1 exPFC2 GABA1 GABA2   MG
##     ASC   1204  705    0     0     0    0      0      0     0     0    0
##     END      0    0  254     0     0    0      0      0     0     0    0
##     exCA1    0    0    0   421     0    0      0      0     0     0    0
##     exCA3    0    0    0     0   745    0      0      0     0     0    0
##     exDG     0    0    0     0     0 1453      0      0     0     0    0
##     exPFC    0    0    0     0     0    0   3104    297     0     0    0
##     GABA     0    0    0     0     0    0      0      0   892   823    0
##     MG       0    0    0     0     0    0      0      0     0     0  389
##     NSC      0    0    0     0     0    0      0      0     0     0    0
##     ODC      0    0    0     0     0    0      0      0     0     0    0
##     OPC      0    0    0     0     0    0      0      0     0     0    0
##     &lt;NA&gt;     0    0    0     0     0    0      0      0     0     0    0
##          Cluster.Name
## Cell_Type  NSC ODC1  OPC Unclassified
##     ASC      0    0    0            0
##     END      0    0    0            0
##     exCA1    0    0    0            0
##     exCA3    0    0    0            0
##     exDG     0    0    0            0
##     exPFC    0    0    0            0
##     GABA     0    0    0            0
##     MG       0    0    0            0
##     NSC    201    0    0            0
##     ODC      0 2965    0            0
##     OPC      0    0  684            0
##     &lt;NA&gt;     0    0    0          826
</code></pre>

<pre><code class="r">smart_table(all_clust_res[,c(&quot;Cell_Type&quot;,&quot;cluster_kmean&quot;)])
</code></pre>

<pre><code>##          cluster_kmean
## Cell_Type    1    2    3    4    5    6    7    8    9   10   11   12   13
##     ASC      1    5    2    2   33    6    1    3   20    4    1    7    0
##     END      2    0  226    0    4    8    0    1    2    1    0    0    0
##     exCA1    0    2    0    1   48  175    1    1    9   86    3   26    0
##     exCA3    0    0    0    0   14    8    0    0    2  683    0    0    0
##     exDG     0    0    0    0    0   75    0    0    5    4    0 1354    0
##     exPFC    5    9    0    2 1301   20    4   13   55   22    2   18    0
##     GABA     3 1139    0    0  217   11    0    0    4    3   13    0    0
##     MG       2    0    0    4    9    7  335    3   11    4    0    2    0
##     NSC      0    0  167    0    0   20    0    0    0    0    0    0    0
##     ODC      3    4    0  794   13    6    2  755 1359    3    0    9    0
##     OPC    640    0    0    1   16    2    0    3    6    0    0    3    0
##     &lt;NA&gt;     5   18    5    7    6  268    4    8   34    7  220   52  164
##          cluster_kmean
## Cell_Type   14   15
##     ASC     15 1809
##     END      8    2
##     exCA1   61    8
##     exCA3   38    0
##     exDG    14    1
##     exPFC 1899   51
##     GABA   324    1
##     MG       4    8
##     NSC      2   12
##     ODC     13    4
##     OPC     10    3
##     &lt;NA&gt;    13   15
</code></pre>

<pre><code class="r">smart_table(all_clust_res[,c(&quot;Cluster.Name&quot;,&quot;cluster_kmean&quot;)])
</code></pre>

<pre><code>##               cluster_kmean
## Cluster.Name      1    2    3    4    5    6    7    8    9   10   11   12
##   ASC1            1    4    0    0   26    4    0    1   11    3    1    4
##   ASC2            0    1    2    2    7    2    1    2    9    1    0    3
##   END             2    0  226    0    4    8    0    1    2    1    0    0
##   exCA1           0    2    0    1   48  175    1    1    9   86    3   26
##   exCA3           0    0    0    0   14    8    0    0    2  683    0    0
##   exDG            0    0    0    0    0   75    0    0    5    4    0 1354
##   exPFC1          5    8    0    1 1117   18    3    1   40   21    2   17
##   exPFC2          0    1    0    1  184    2    1   12   15    1    0    1
##   GABA1           3  359    0    0  202    9    0    0    4    3    8    0
##   GABA2           0  780    0    0   15    2    0    0    0    0    5    0
##   MG              2    0    0    4    9    7  335    3   11    4    0    2
##   NSC             0    0  167    0    0   20    0    0    0    0    0    0
##   ODC1            3    4    0  794   13    6    2  755 1359    3    0    9
##   OPC           640    0    0    1   16    2    0    3    6    0    0    3
##   Unclassified    5   18    5    7    6  268    4    8   34    7  220   52
##               cluster_kmean
## Cluster.Name     13   14   15
##   ASC1            0    5 1144
##   ASC2            0   10  665
##   END             0    8    2
##   exCA1           0   61    8
##   exCA3           0   38    0
##   exDG            0   14    1
##   exPFC1          0 1835   36
##   exPFC2          0   64   15
##   GABA1           0  304    0
##   GABA2           0   20    1
##   MG              0    4    8
##   NSC             0    2   12
##   ODC1            0   13    4
##   OPC             0   10    3
##   Unclassified  164   13   15
</code></pre>

<pre><code class="r">smart_table(all_clust_res[,c(&quot;cluster_kmean&quot;,&quot;sc3_15_clusters&quot;)])
</code></pre>

<pre><code>##              sc3_15_clusters
## cluster_kmean    1    2    3    4    5    6    7    8    9   10   11   12
##            1     1    6  562   10   11    0   10    0    0    6   36    0
##            2    21    1    0   61    1    8    7    1    0 1028    1    0
##            3     0   33    0  333    0    0    2    0    4    5    0   20
##            4     0    1    0    0  631    0  121    0    0    2    0    0
##            5  1075    6    4   12    0    0   31    6    7  199    0    0
##            6     6    4   31   56    3    0    2   15  272   20    4    0
##            7     0    5    4  294    8    0    8    0    2    1    1   20
##            8     0    1    2    0  224    0  497    0    1    1    0    0
##            9     2    3    2    2  470    0  913    1    7    6    1    0
##            10   43    1    0    0    0    0    0   61   10   13    0    0
##            11    3    0    0  194    1    0    0    0    2   15    1   15
##            12    9    0   77    0    0    0    0    0 1361    1    0    0
##            13    0    0    0  149    0    0    0    0    0    0    0   15
##            14 1374   11    5   36    0    0   47   14   18  303    0    1
##            15    5  654    8    7    4    0   17    1   76   61    2    2
##              sc3_15_clusters
## cluster_kmean   13   14   15
##            1     7    2   10
##            2    45    1    2
##            3     0    0    3
##            4    50    2    4
##            5   287   31    3
##            6     1  188    4
##            7     0    1    3
##            8    17    3   41
##            9    27    2   71
##            10    1  686    2
##            11    2    6    0
##            12    1   22    0
##            13    0    0    0
##            14  526   65    1
##            15    2    6 1069
</code></pre>

<pre><code class="r">smart_table(all_clust_res[,c(&quot;Cluster.Name&quot;,&quot;sc3_15_clusters&quot;)])
</code></pre>

<pre><code>##               sc3_15_clusters
## Cluster.Name      1    2    3    4    5    6    7    8    9   10   11   12
##   ASC1           28  234    7    3    8    0   13    1   67   29    1    0
##   ASC2            4  391    2    5    5    0    7    0   16   36    0    1
##   END             3    3    2  221    0    0    3    0    2    1    0   15
##   exCA1          38    2    1    5    5    0    5   28   32   13    1    0
##   exCA3          52    2    0    0    0    0    1   53    8   11    0    0
##   exDG           10    0   82    0    0    0    3    0 1337    1    0    0
##   exPFC1       2246   25    7   18   18    0   63   11   22   78    0    0
##   exPFC2         56    7    0    1    6    0   41    1    3    5    0    0
##   GABA1          34    2    2   43    1    1   14    1    0  721    1    0
##   GABA2          24    1    0   46    0    7    8    1    0  698    1    0
##   MG              5    6    5  297   16    0   12    0    6    7    2   20
##   NSC             0   37    0  141    0    0    0    0    6    4    0    7
##   ODC1           15    4    3    0 1252    0 1451    2   18   13    0    0
##   OPC             9    6  558    7   14    0   13    0    4   11   38    0
##   Unclassified   15    6   26  367   28    0   21    1  239   33    2   30
##               sc3_15_clusters
## Cluster.Name     13   14   15
##   ASC1            8    6  799
##   ASC2            5    4  229
##   END             1    2    1
##   exCA1          11  276    4
##   exCA3           4  613    1
##   exDG            4   15    1
##   exPFC1        549   38   29
##   exPFC2        166    5    6
##   GABA1          68    3    1
##   GABA2          35    0    2
##   MG              5    2    6
##   NSC             0    0    6
##   ODC1           96    6  105
##   OPC            11    1   12
##   Unclassified    3   44   11
</code></pre>

<pre><code class="r">smart_table(all_clust_res[,c(&quot;cluster_kmean&quot;,&quot;sc3_10_clusters&quot;)])
</code></pre>

<pre><code>##              sc3_10_clusters
## cluster_kmean    1    2    3    4    5    6    7    8    9   10
##            1     0  575    4    0    5   23   15    0    0   39
##            2    29   64 1040    4   38    0    2    0    0    0
##            3     0  242    1    0    0    6   73   60    4   14
##            4     0    1    1    0   53  754    0    0    0    2
##            5    11   17  220  291 1079    2    5    0   10   26
##            6     0   59   23    1   10    3   13   31  288  178
##            7     0  298    0    0    0   18    7    0    2   22
##            8     0    1    1    0   56  724    2    1    1    1
##            9     0    6    6    0   90 1390    8    0    4    3
##            10    0    0   13    2   79    2    2    0   67  652
##            11    0  190   17    2    6    0    0    0    3   21
##            12    0    1    1    2   13    1    0   77 1358   18
##            13    0  149    0    0    0    0    0    0    0   15
##            14   25   53  344  445 1464    1    1    0   19   49
##            15    0   23    5    0   10  154 1714    0    5    3
</code></pre>

<pre><code class="r">smart_table(all_clust_res[,c(&quot;Cluster.Name&quot;,&quot;sc3_10_clusters&quot;)])
</code></pre>

<pre><code>##               sc3_10_clusters
## Cluster.Name      1    2    3    4    5    6    7    8    9   10
##   ASC1            0   13    6    7   31  106 1031    0    6    4
##   ASC2            0   11    3    1   11   57  613    0    8    1
##   END             0  223    0    0    4    4    5    0    1   17
##   exCA1           0    9   12    9   56   11    6    1   62  255
##   exCA3           0    0   12    6   83    4    0    0   61  579
##   exDG            0    1    2    3   18    2    1   82 1332   12
##   exPFC1          3   34  113  546 2275   44   37    0   20   32
##   exPFC2          0    1   28  152   73   29   11    0    3    0
##   GABA1          37   49  733    5   61    4    1    0    0    2
##   GABA2          25   50  703    4   39    0    2    0    0    0
##   MG              0  304    5    1    9   29   11    1    6   23
##   NSC             0   42    1    0    0    4   86   61    6    1
##   ODC1            0    5   14    9  204 2708    7    1   13    4
##   OPC             0  569    8    2   17   29   15    0    3   41
##   Unclassified    0  368   36    2   22   47   16   23  240   72
</code></pre>

<pre><code class="r"># Plot our TSNE results colored by our kmeans result
gp1 = ggplot(all_clust_res, aes(X1,X2,col=cluster_kmean)) + 
  geom_point(size=0.2,alpha=0.6) + theme_classic() + 
  # scale_color_manual(values=cols) + 
  guides(color = guide_legend(override.aes = list(size=3)))
gp1
</code></pre>

<p><img src="figure/unnamed-chunk-18-2.png" alt="plot of chunk unnamed-chunk-18"></p>

<pre><code class="r"># Plot our TSNE results but colored by paper supplement&#39;s cell types
gp1 = ggplot(all_clust_res, aes(X1,X2,col=Cell_Type)) + 
  geom_point(size=0.2,alpha=0.6) + theme_classic() + 
  # scale_color_manual(values=cols) + 
  guides(color = guide_legend(override.aes = list(size=3)))
gp1
</code></pre>

<p><img src="figure/unnamed-chunk-18-3.png" alt="plot of chunk unnamed-chunk-18"></p>

<pre><code class="r"># Plot our TSNE results but colored by paper supplement&#39;s clusters
gp1 = ggplot(all_clust_res, aes(X1,X2,col=Cluster.Name)) + 
  geom_point(size=0.2,alpha=0.6) + theme_classic() + 
  # scale_color_manual(values=cols) + 
  guides(color = guide_legend(override.aes = list(size=3)))
gp1
</code></pre>

<p><img src="figure/unnamed-chunk-18-4.png" alt="plot of chunk unnamed-chunk-18"></p>

<pre><code class="r"># Plot paper&#39;s TSNE plot with their cell types
paper_tsne = smart_RT(file.path(data_dir,&quot;GTEx_droncseq_hip_pcf.tsne.txt&quot;),sep=&#39;\t&#39;,header=TRUE,row.names=1)
paper_tsne$Cell.ID = rownames(paper_tsne)
rownames(paper_tsne) = NULL
all_clust_res = smart_merge(all_clust_res,paper_tsne)
gp1 = ggplot(all_clust_res, aes(tSNE_1,tSNE_2,col=Cell_Type)) + 
  geom_point(size=0.2,alpha=0.6) + theme_classic() + 
  # scale_color_manual(values=cols) + 
  guides(color = guide_legend(override.aes = list(size=3)))
gp1
</code></pre>

<p><img src="figure/unnamed-chunk-18-5.png" alt="plot of chunk unnamed-chunk-18"></p>

<pre><code class="r"># Plot paper&#39;s TSNE plot with our clusters
gp1 = ggplot(all_clust_res, aes(tSNE_1,tSNE_2,col=cluster_kmean)) + 
  geom_point(size=0.2,alpha=0.6) + theme_classic() + 
  # scale_color_manual(values=cols) + 
  guides(color = guide_legend(override.aes = list(size=3)))
gp1
</code></pre>

<p><img src="figure/unnamed-chunk-18-6.png" alt="plot of chunk unnamed-chunk-18"></p>

<pre><code class="r"># Plot our TSNE plot with our SC3 15 clusters
gp1 = ggplot(all_clust_res, aes(X1,X2,col=factor(sc3_15_clusters))) + 
  geom_point(size=0.2,alpha=0.6) + theme_classic() + 
  # scale_color_manual(values=cols) + 
  guides(color = guide_legend(override.aes = list(size=3)))
gp1
</code></pre>

<p><img src="figure/unnamed-chunk-18-7.png" alt="plot of chunk unnamed-chunk-18"></p>

<pre><code class="r"># Plot our TSNE plot with our SC3 10 clusters
gp1 = ggplot(all_clust_res, aes(X1,X2,col=factor(sc3_10_clusters))) + 
  geom_point(size=0.2,alpha=0.6) + theme_classic() + 
  # scale_color_manual(values=cols) + 
  guides(color = guide_legend(override.aes = list(size=3)))
gp1
</code></pre>

<p><img src="figure/unnamed-chunk-18-8.png" alt="plot of chunk unnamed-chunk-18"></p>

<pre><code class="r"># Save all three clustering results
all_clust_res_fn = file.path(data_dir,&quot;all_clust_res.rds&quot;)
saveRDS(all_clust_res,all_clust_res_fn)
</code></pre>

<h1>Session information</h1>

<pre><code class="r">sessionInfo()
</code></pre>

<pre><code>## R version 3.5.1 (2018-07-02)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Red Hat Enterprise Linux Server 7.5 (Maipo)
## 
## Matrix products: default
## BLAS: /nas/longleaf/home/pllittle/downloads/R-3.5.1/lib64/R/lib/libRblas.so
## LAPACK: /nas/longleaf/home/pllittle/downloads/R-3.5.1/lib64/R/lib/libRlapack.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] SC3_1.8.0                   Rtsne_0.13                 
##  [3] svd_0.4.1                   limma_3.36.3               
##  [5] scran_1.8.4                 scater_1.8.4               
##  [7] ggplot2_3.0.0               biomaRt_2.36.1             
##  [9] DropletUtils_1.0.3          SingleCellExperiment_1.2.0 
## [11] SummarizedExperiment_1.10.1 DelayedArray_0.6.5         
## [13] BiocParallel_1.14.2         matrixStats_0.54.0         
## [15] Biobase_2.40.0              GenomicRanges_1.32.6       
## [17] GenomeInfoDb_1.16.0         IRanges_2.14.11            
## [19] S4Vectors_0.18.3            BiocGenerics_0.26.0        
## [21] BiocInstaller_1.30.0        markdown_0.8               
## [23] knitr_1.20                 
## 
## loaded via a namespace (and not attached):
##   [1] ggbeeswarm_0.6.0         colorspace_1.3-2        
##   [3] rjson_0.2.20             class_7.3-14            
##   [5] dynamicTreeCut_1.63-1    XVector_0.20.0          
##   [7] DT_0.4                   bit64_0.9-7             
##   [9] mvtnorm_1.0-8            AnnotationDbi_1.42.1    
##  [11] codetools_0.2-15         tximport_1.8.0          
##  [13] doParallel_1.0.11        robustbase_0.93-2       
##  [15] cluster_2.0.7-1          pheatmap_1.0.10         
##  [17] shinydashboard_0.7.0     shiny_1.1.0             
##  [19] rrcov_1.4-4              compiler_3.5.1          
##  [21] httr_1.3.1               assertthat_0.2.0        
##  [23] Matrix_1.2-14            lazyeval_0.2.1          
##  [25] later_0.7.3              htmltools_0.3.6         
##  [27] prettyunits_1.0.2        tools_3.5.1             
##  [29] bindrcpp_0.2.2           igraph_1.2.2            
##  [31] gtable_0.2.0             glue_1.3.0              
##  [33] GenomeInfoDbData_1.1.0   reshape2_1.4.3          
##  [35] dplyr_0.7.6              doRNG_1.7.1             
##  [37] Rcpp_0.12.18             gdata_2.18.0            
##  [39] iterators_1.0.10         DelayedMatrixStats_1.2.0
##  [41] stringr_1.3.1            mime_0.5                
##  [43] irlba_2.3.2              rngtools_1.3.1          
##  [45] gtools_3.8.1             WriteXLS_4.0.0          
##  [47] statmod_1.4.30           XML_3.98-1.16           
##  [49] DEoptimR_1.0-8           edgeR_3.22.3            
##  [51] zlibbioc_1.26.0          scales_1.0.0            
##  [53] hms_0.4.2                promises_1.0.1          
##  [55] rhdf5_2.24.0             RColorBrewer_1.1-2      
##  [57] curl_3.2                 memoise_1.1.0           
##  [59] gridExtra_2.3            pkgmaker_0.27           
##  [61] stringi_1.2.4            RSQLite_2.1.1           
##  [63] highr_0.7                pcaPP_1.9-73            
##  [65] foreach_1.4.4            e1071_1.7-0             
##  [67] caTools_1.17.1.1         bibtex_0.4.2            
##  [69] rlang_0.2.2              pkgconfig_2.0.2         
##  [71] bitops_1.0-6             evaluate_0.11           
##  [73] lattice_0.20-35          ROCR_1.0-7              
##  [75] purrr_0.2.5              Rhdf5lib_1.2.1          
##  [77] bindr_0.1.1              htmlwidgets_1.2         
##  [79] labeling_0.3             bit_1.1-14              
##  [81] tidyselect_0.2.4         plyr_1.8.4              
##  [83] magrittr_1.5             R6_2.2.2                
##  [85] gplots_3.0.1             DBI_1.0.0               
##  [87] pillar_1.3.0             withr_2.1.2             
##  [89] RCurl_1.95-4.11          tibble_1.4.2            
##  [91] crayon_1.3.4             KernSmooth_2.23-15      
##  [93] viridis_0.5.1            progress_1.2.0          
##  [95] locfit_1.5-9.1           grid_3.5.1              
##  [97] data.table_1.11.4        blob_1.1.1              
##  [99] FNN_1.1.2.1              digest_0.6.16           
## [101] xtable_1.8-3             httpuv_1.4.5            
## [103] munsell_0.5.0            registry_0.5            
## [105] beeswarm_0.2.3           viridisLite_0.3.0       
## [107] vipor_0.4.5
</code></pre>

<h1>Reference</h1>

</body>

</html>
